// @dnd-kit/accessibility/dist/accessibility.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);

var hiddenStyles = {
  display: 'none'
};
function HiddenText(_ref) {
  var id = _ref.id,
      value = _ref.value;
  return React__default.createElement("div", {
    id: id,
    style: hiddenStyles
  }, value);
}

var visuallyHidden = {
  position: 'absolute',
  width: 1,
  height: 1,
  margin: -1,
  border: 0,
  padding: 0,
  overflow: 'hidden',
  clip: 'rect(0 0 0 0)',
  clipPath: 'inset(100%)',
  whiteSpace: 'nowrap'
};
function LiveRegion(_ref) {
  var id = _ref.id,
      announcement = _ref.announcement;
  return React__default.createElement("div", {
    id: id,
    style: visuallyHidden,
    role: "status",
    "aria-live": "assertive",
    "aria-atomic": true
  }, announcement);
}

function useAnnouncement() {
  var _useState = React.useState(''),
      announcement = _useState[0],
      setAnnouncement = _useState[1];

  return {
    announce: setAnnouncement,
    announcement: announcement
  };
}

exports.HiddenText = HiddenText;
exports.LiveRegion = LiveRegion;
exports.useAnnouncement = useAnnouncement;
