// @element/react-drawer/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime-corejs3/helpers/extends');
var drawer = require('@material/drawer');
var clsx = require('clsx');
var React = require('react');
var reactUse = require('react-use');
var PropTypes = require('prop-types');
var reactCommon = require('@element/react-common');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefault(_extends);
var clsx__default = /*#__PURE__*/_interopDefault(clsx);
var React__default = /*#__PURE__*/_interopDefault(React);
var PropTypes__default = /*#__PURE__*/_interopDefault(PropTypes);

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const DrawerPropTypes = {
  /**
   * Positions the modal lower, below the TopAppBar.
   *
   * Defaults to **null**.
   */
  belowTopAppBar: PropTypes__default["default"].bool,
  /**
   * Content to be rendered inside the Drawer
   *
   * Defaults to **null**.
   */
  children: PropTypes__default["default"].node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes__default["default"].string,
  /**
   * @deprecated _Use `variant='dismissible'` instead._
   *
   * Enables a dismissible drawer.
   *
   * Defaults to **undefined**.
   */
  dismissible: reactCommon.deprecatedProp(PropTypes__default["default"].bool, "Drawer", 'Use `variant="dismissible"` instead.'),
  /**
   * @deprecated _Use `variant='modal'` instead._
   *
   * Enabled a modal drawer.
   *
   * Defaults to **undefined**.
   */
  modal: reactCommon.deprecatedProp(PropTypes__default["default"].bool, "Drawer", 'Use `variant="modal"` instead.'),
  /**
   * Fired when drawer is closed.
   *
   * Defaults to **null**.
   */
  onClose: PropTypes__default["default"].func,
  /**
   * Fired when drawer is opened.
   *
   * Defaults to **null**.
   */
  onOpen: PropTypes__default["default"].func,
  /**
   * Displays the drawer when `variant` is either `dismissible` or modal`
   *
   * Defaults to **null**.
   */
  open: PropTypes__default["default"].bool,
  /**
   * Defines the type of drawer. By default the drawer will be always shown, a permanent drawer. If defined, either `modal` or `dismissible`
   *
   * Defaults to **empty string**.
   */
  variant: PropTypes__default["default"].oneOf(["", "modal", "dismissible"])
};
const DrawerDefaultProps = {
  belowTopAppBar: null,
  children: null,
  className: undefined,
  dismissible: undefined,
  modal: undefined,
  open: null,
  variant: ""
};

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const DrawerAppContentPropTypes = {
  /**
   * Positions the modal lower, below the TopAppBar.
   *
   * Defaults to **null**.
   */
  belowTopAppBar: PropTypes__default["default"].bool,
  /**
   * Content to be rendered inside the DrawerAppContent
   *
   * Defaults to **null**.
   */
  children: PropTypes__default["default"].node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes__default["default"].string
};
const DrawerAppContentDefaultProps = {
  belowTopAppBar: null,
  children: null,
  className: undefined
};

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const DrawerContentPropTypes = {
  /**
   * Content to be rendered inside the drawer. Most commonly a list, but accepts any valid markup.
   *
   * Defaults to **null**.
   */
  children: PropTypes__default["default"].node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes__default["default"].string
};
const DrawerContentDefaultProps = {
  children: null,
  className: undefined
};

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const DrawerHeaderPropTypes = {
  /**
   * Most commonly an avatar style image, but accepts any valid markup.
   *
   * Defaults to **null**.
   */
  avatar: PropTypes__default["default"].node,
  /**
   * Most commonly a text or a Typography component, but accepts any valid markup.
   *
   * Defaults to **null**.
   */
  children: PropTypes__default["default"].node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes__default["default"].string,
  /**
   * Subtitle of Drawer Header
   *
   * Defaults to **empty string**.
   */
  subtitle: PropTypes__default["default"].string,
  /**
   * Title of Drawer Header
   *
   * Defaults to **empty string**.
   */
  title: PropTypes__default["default"].string
};
const DrawerHeaderDefaultProps = {
  avatar: null,
  children: null,
  className: undefined,
  subtitle: "",
  title: ""
};

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const DrawerScrimPropTypes = {};
const DrawerScrimDefaultProps = {};

const DrawerScrim = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    className,
    ...customProps
  } = props;
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    ref: externalRef
  }, customProps, {
    className: clsx__default["default"]('lmnt mdc-drawer-scrim', className)
  }));
});
DrawerScrim.displayName = 'DrawerScrim';
DrawerScrim.propTypes = DrawerScrimPropTypes;
DrawerScrim.defaultProps = DrawerScrimDefaultProps;
var DrawerScrim$1 = /*#__PURE__*/React__default["default"].memo(DrawerScrim);

const Drawer = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    belowTopAppBar,
    children,
    className,
    dismissible,
    modal,
    onClose,
    onOpen,
    open,
    variant,
    ...customProps
  } = props;
  const [ref, setRef] = React.useState(null); // Initialize to null so that useEvent doesn't initially listen to the global window
  const [mdcDrawer, setMdcDrawer] = React.useState(null);
  const [isInitiallyOpen] = React.useState(() => !!open);
  const drawerRef = React.useRef();
  React.useImperativeHandle(externalRef, () => drawerRef.current, []);
  const isDismissible = variant === 'dismissible' || dismissible;
  const isModal = variant === 'modal' || modal;
  const isCloseable = isDismissible || isModal;
  React.useEffect(() => {
    const drawer$1 = ref && isCloseable ? drawer.MDCDrawer.attachTo(ref) : null;
    setMdcDrawer(drawer$1);
    return () => {
      drawer$1 === null || drawer$1 === void 0 ? void 0 : drawer$1.destroy();
    };
  }, [isCloseable, ref]);
  reactUse.useEvent('MDCDrawer:opened', onOpen, ref);
  reactUse.useEvent('MDCDrawer:closed', onClose, ref);
  React.useEffect(() => {
    if (mdcDrawer) {
      mdcDrawer.open = open;
    }
  }, [mdcDrawer, open]);
  const classes = React.useMemo(() => clsx__default["default"]('lmnt', 'mdc-drawer', !variant && 'lmnt-drawer--permanent', isInitiallyOpen && 'mdc-drawer--open', isDismissible && 'mdc-drawer--dismissible', isModal && 'mdc-drawer--modal', belowTopAppBar && 'mdc-top-app-bar--fixed-adjust', className), [belowTopAppBar, className, isDismissible, isInitiallyOpen, isModal, variant]);
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement("aside", _extends__default["default"]({
    ref: setRef,
    className: classes
  }, customProps), children), isModal && /*#__PURE__*/React__default["default"].createElement(DrawerScrim$1, null));
});
Drawer.displayName = 'Drawer';
Drawer.propTypes = DrawerPropTypes;
Drawer.defaultProps = DrawerDefaultProps;
var Drawer$1 = /*#__PURE__*/React__default["default"].memo(Drawer);

const DrawerAppContent = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    className,
    belowTopAppBar,
    children,
    ...customProps
  } = props;
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    ref: externalRef,
    className: clsx__default["default"]('lmnt', 'mdc-drawer-app-content', belowTopAppBar && 'mdc-top-app-bar--fixed-adjust', className)
  }, customProps), children);
});
DrawerAppContent.displayName = 'DrawerAppContent';
DrawerAppContent.propTypes = DrawerAppContentPropTypes;
DrawerAppContent.defaultProps = DrawerAppContentDefaultProps;
var DrawerAppContent$1 = /*#__PURE__*/React__default["default"].memo(DrawerAppContent);

const DrawerContent = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    children,
    className,
    ...customProps
  } = props;
  return /*#__PURE__*/React__default["default"].createElement("nav", _extends__default["default"]({
    ref: externalRef,
    className: clsx__default["default"]('lmnt', 'mdc-drawer__content', className)
  }, customProps), children);
});
DrawerContent.displayName = 'DrawerContent';
DrawerContent.propTypes = DrawerContentPropTypes;
DrawerContent.defaultProps = DrawerContentDefaultProps;
var DrawerContent$1 = /*#__PURE__*/React__default["default"].memo(DrawerContent);

const DrawerHeader = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    children,
    avatar,
    className,
    title,
    subtitle,
    ...customProps
  } = props;
  let avatarRendered = null;
  if (avatar) {
    avatarRendered = /*#__PURE__*/React__default["default"].cloneElement(avatar, {
      className: 'mdc-drawer__header-avatar'
    });
  }
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    ref: externalRef,
    className: clsx__default["default"]('lmnt', 'mdc-drawer__header', className)
  }, customProps), avatarRendered && /*#__PURE__*/React__default["default"].createElement("div", null, avatarRendered), title && /*#__PURE__*/React__default["default"].createElement("div", {
    className: "mdc-drawer__title"
  }, title), subtitle && /*#__PURE__*/React__default["default"].createElement("div", {
    className: "mdc-drawer__subtitle"
  }, subtitle), children);
});
DrawerHeader.displayName = 'DrawerHeader';
DrawerHeader.propTypes = DrawerHeaderPropTypes;
DrawerHeader.defaultProps = DrawerHeaderDefaultProps;
var DrawerHeader$1 = /*#__PURE__*/React__default["default"].memo(DrawerHeader);

exports.Drawer = Drawer$1;
exports.DrawerAppContent = DrawerAppContent$1;
exports.DrawerContent = DrawerContent$1;
exports.DrawerHeader = DrawerHeader$1;
exports.DrawerScrim = DrawerScrim$1;
