// @monsantoit/profile-client/lib/deprecated/globalSettings.js
"use strict";

var _ = require('../lodash');
var browser = {
  window: function (_window) {
    function window() {
      return _window.apply(this, arguments);
    }
    window.toString = function () {
      return _window.toString();
    };
    return window;
  }(function () {
    if (typeof window === 'undefined') {
      return {};
    }
    return window;
  }),
  document: function (_document) {
    function document() {
      return _document.apply(this, arguments);
    }
    document.toString = function () {
      return _document.toString();
    };
    return document;
  }(function () {
    if (typeof document === 'undefined') {
      return {};
    }
    return document;
  }),
  namespace: function namespace() {
    var namespace = browser.window().velocity || browser.window().phoenix;
    if (namespace != null) {
      return namespace;
    }
    console.warn('Could not find the namespace for profile client. Are you using @monsantoit/service-bindings middleware?');
    return {};
  },
  serviceBindings: function serviceBindings() {
    return browser.namespace().serviceBindings || {};
  },
  localDevelopment: function localDevelopment() {
    return !!browser.namespace().localDevelopment;
  },
  localOcelot: function localOcelot() {
    return !!browser.namespace().localOcelot;
  }
};

// profile-client settings, derived from browser globals
var settings = {
  debuggerEnabled: function debuggerEnabled() {
    return !!browser.namespace().debugProfileClient;
  },
  refreshEnabled: function refreshEnabled() {
    if (_.has(browser.namespace(), 'refreshEnabled')) {
      return browser.namespace().refreshEnabled;
    }
    return true;
  },
  authEnabled: function authEnabled() {
    var notLocalDevelopment = browser.localDevelopment() === false;
    var localOcelot = browser.localOcelot() === true;
    return notLocalDevelopment || localOcelot;
  },
  retryInterval: function retryInterval() {
    return 1000 * 5;
  },
  tokenFromCookie: function tokenFromCookie(name) {
    var cookiePairs = (browser.document().cookie || '').split(';').map(function (cookie) {
      return cookie.trim().split('=');
    });
    var cookieMap = _.fromPairs(cookiePairs) || {};
    if (name != null && cookieMap[name]) {
      return cookieMap[name];
    }
    return null;
  }
};
module.exports = {
  browser: browser,
  settings: settings
};
