// @element/react-snackbar/lib/index.es.js
import _extends from '@babel/runtime-corejs3/helpers/extends';
import React, { forwardRef, useRef, useImperativeHandle, useEffect } from 'react';
import clsx from 'clsx';
import { MDCSnackbar } from '@material/snackbar';
import { edsWarning } from '@element/react-common';
import PropTypes from 'prop-types';

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const SnackbarPropTypes = {
  /**
   * Optional action button label. If set, an action button will be shown.
   *
   * Defaults to **null**.
   */
  actionLabel: PropTypes.string,
  /**
   * Most commonly a string, but will accept any valid components or markup. If `label` and `children` are both set, `label` will take priority.
   *
   * Defaults to **null**.
   */
  children: PropTypes.node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes.string,
  /**
   * Sets whether the Snackbar closes when it is focused and the user presses the ESC key.
   *
   * Defaults to **true**.
   */
  closeOnEscape: PropTypes.bool,
  /**
   * Optional dismiss button label. If set, an button button will be shown.
   *
   * Defaults to **null**.
   */
  dismissLabel: PropTypes.string,
  /**
   * The snackbar text. If `label` and `children` are both set, `label` will take priority.
   *
   * Defaults to **null**.
   */
  label: PropTypes.string,
  /**
   * Positions the Snackbar on the leading edge of the screen (left in LTR, right in RTL) instead of centered.
   *
   * Defaults to **false**.
   */
  leading: PropTypes.bool,
  /**
   * Fired when the Snackbar action button is clicked.
   *
   * Defaults to **null**.
   */
  onClick: PropTypes.func,
  /**
   * Fired when the Snackbar finishes its closing animation.
   *
   * Defaults to **null**.
   */
  onClose: PropTypes.func,
  /**
   * Fired when the Snackbar begins it's closing animation.
   *
   * Defaults to **null**.
   */
  onClosing: PropTypes.func,
  /**
   * Fired when the Snackbar finishes its opening animation.
   *
   * Defaults to **null**.
   */
  onOpen: PropTypes.func,
  /**
   * Fired when the Snackbar begins it's opening animation.
   *
   * Defaults to **null**.
   */
  onOpening: PropTypes.func,
  /**
   * Indicates that the Snackbar is open and visible.
   *
   * Defaults to **false**.
   */
  open: PropTypes.bool,
  /**
   * Positions the action button/icon below the label instead of alongside it.
   *
   * Defaults to **false**.
   */
  stacked: PropTypes.bool,
  /**
   * The time the snackbar will remain visible. The value must be between `4000` and `10000` or set to `-1` to disable the timeout.
   *
   * Defaults to **5000**.
   */
  timeout: PropTypes.number
};
const SnackbarDefaultProps = {
  actionLabel: null,
  children: null,
  className: undefined,
  closeOnEscape: true,
  dismissLabel: null,
  label: null,
  leading: false,
  open: false,
  stacked: false,
  timeout: 5000
};

const Snackbar = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    actionLabel,
    children,
    className,
    closeOnEscape,
    dismissLabel,
    label,
    leading,
    onClick,
    onClose,
    onClosing,
    onOpen,
    onOpening,
    open,
    stacked,
    timeout,
    ...customProps
  } = props;
  const mdcSnackbar = useRef();
  const ref = useRef();
  useImperativeHandle(externalRef, () => ref.current, []);
  useEffect(() => {
    mdcSnackbar.current = MDCSnackbar.attachTo(ref.current);
    return () => {
      mdcSnackbar.current.destroy();
    };
  }, []);
  useEffect(() => {
    mdcSnackbar.current.listen('MDCSnackbar:closed', onClose);
    return () => {
      mdcSnackbar.current.unlisten('MDCSnackbar:closed', onClose);
    };
  }, [onClose]);
  useEffect(() => {
    mdcSnackbar.current.listen('MDCSnackbar:closed', onClosing);
    return () => {
      mdcSnackbar.current.unlisten('MDCSnackbar:closed', onClosing);
    };
  }, [onClosing]);
  useEffect(() => {
    mdcSnackbar.current.listen('MDCSnackbar:closed', onOpen);
    return () => {
      mdcSnackbar.current.unlisten('MDCSnackbar:closed', onOpen);
    };
  }, [onOpen]);
  useEffect(() => {
    mdcSnackbar.current.listen('MDCSnackbar:closed', onOpening);
    return () => {
      mdcSnackbar.current.unlisten('MDCSnackbar:closed', onOpening);
    };
  }, [onOpening]);
  useEffect(() => {
    if (actionLabel) mdcSnackbar.current.actionButtonText = actionLabel;
  }, [actionLabel]);
  useEffect(() => {
    if (label) mdcSnackbar.current.labelText = label;
  }, [label]);
  useEffect(() => {
    if (open) mdcSnackbar.current.open();
    if (!open) mdcSnackbar.current.close();
  }, [open]);
  useEffect(() => {
    mdcSnackbar.current.closeOnEscape = closeOnEscape;
  }, [closeOnEscape]);
  useEffect(() => {
    if (timeout && (timeout === -1 || timeout >= 4000 && timeout <= 10000)) {
      mdcSnackbar.current.timeoutMs = timeout;
    } else {
      edsWarning('Timeout must be between `4000` and `10000` or set to `-1` to disable.');
    }
  }, [timeout]);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx('lmnt', 'mdc-snackbar', 'lmnt-snackbar', leading && 'mdc-snackbar--leading', stacked && 'mdc-snackbar--stacked', className)
  }, customProps), /*#__PURE__*/React.createElement("div", {
    className: "mdc-snackbar__surface",
    role: "status",
    "aria-relevant": "additions"
  }, /*#__PURE__*/React.createElement("div", {
    className: "mdc-snackbar__label",
    "aria-atomic": "false"
  }, children), (actionLabel || dismissLabel) && /*#__PURE__*/React.createElement("div", {
    className: "mdc-snackbar__actions",
    "aria-atomic": "true"
  }, actionLabel && /*#__PURE__*/React.createElement("button", {
    tabIndex: 0,
    className: "mdc-button mdc-snackbar__action",
    type: "button",
    onClick: onClick
  }, /*#__PURE__*/React.createElement("div", {
    className: "mdc-button__ripple"
  }), /*#__PURE__*/React.createElement("span", {
    className: "mdc-button__label"
  }, actionLabel)), dismissLabel && /*#__PURE__*/React.createElement("button", {
    tabIndex: 0,
    className: "mdc-button mdc-snackbar__action",
    type: "button"
  }, /*#__PURE__*/React.createElement("div", {
    className: "mdc-button__ripple"
  }), /*#__PURE__*/React.createElement("span", {
    className: "mdc-button__label"
  }, dismissLabel)))));
});
Snackbar.displayName = 'Snackbar';
Snackbar.propTypes = SnackbarPropTypes;
Snackbar.defaultProps = SnackbarDefaultProps;

export { Snackbar };
