// @element/themes/lib/tokens/asgrow.theme.js
export const primaryMap = {
    'primary-50': '#fefef7',
    'primary-100': '#fafaf0',
    'primary-200': '#f7f7e9',
    'primary-300': '#f2efe1',
    'primary-400': '#e8e6cc',
    'primary-500': '#d2cbaa',
    'primary-600': '#b8b597',
    'primary-700': '#989174',
    'primary-800': '#7f7569',
    'primary-900': '#535044'
};

export const secondaryMap = {
    'secondary-50': '#e6f9ed',
    'secondary-100': '#aee9c6',
    'secondary-200': '#59d38a',
    'secondary-300': '#03bd5e',
    'secondary-400': '#00a651',
    'secondary-500': '#00974a',
    'secondary-600': '#008331',
    'secondary-700': '#006f29',
    'secondary-800': '#006027',
    'secondary-900': '#003c18'
};

export const dangerMap = {
    'danger-50': '#ffeaed',
    'danger-100': '#ffccce',
    'danger-200': '#ee9894',
    'danger-300': '#e3716a',
    'danger-400': '#ec5145',
    'danger-500': '#ef4227',
    'danger-600': '#e13727',
    'danger-700': '#cf2d22',
    'danger-800': '#c2261b',
    'danger-900': '#b3190d'
};

export const successMap = {
    'success-50': '#E9FFF8',
    'success-100': '#B9FFEA',
    'success-200': '#87FFDB',
    'success-300': '#30F3B9',
    'success-400': '#1DE1A7',
    'success-500': '#14C38F',
    'success-600': '#0EA074',
    'success-700': '#0A805D',
    'success-800': '#085C43',
    'success-900': '#05392A'
};

export const ancillaryMap = {
    'ancillary-100': '#a74f28'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#240F67',
    'dataVizComparative-100': '#2D72C4',
    'dataVizComparative-200': '#89C9B7',
    'dataVizComparative-300': '#785E9F',
    'dataVizComparative-400': '#3B8EC8',
    'dataVizComparative-500': '#B7DBBC',
    'dataVizComparative-600': '#4153AF',
    'dataVizComparative-700': '#47A3C4',
    'dataVizComparative-800': '#DFECCB',
    'dataVizComparative-900': '#6332AD'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#261643',
    'dataVizSequential1-100': '#422575',
    'dataVizSequential1-200': '#6540AA',
    'dataVizSequential1-300': '#7E52D0',
    'dataVizSequential1-400': '#9A75DD',
    'dataVizSequential1-500': '#AF8BEF',
    'dataVizSequential1-600': '#C6A6FF',
    'dataVizSequential1-700': '#D8C1FF',
    'dataVizSequential1-800': '#E8DBFF',
    'dataVizSequential1-900': '#F5EFFF'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#052D44',
    'dataVizSequential2-100': '#08517B',
    'dataVizSequential2-200': '#1774AA',
    'dataVizSequential2-300': '#319BD8',
    'dataVizSequential2-400': '#40BAFF',
    'dataVizSequential2-500': '#75CDFF',
    'dataVizSequential2-600': '#97E0FF',
    'dataVizSequential2-700': '#A9F0FF',
    'dataVizSequential2-800': '#C3FBFF',
    'dataVizSequential2-900': '#DDFFFD'
};
