// @element/themes/lib/tokens/aspirincardio.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    dangerMap,
    successMap,
    lmntThemeDanger,
    lmntThemeDangerVariant,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#01866A';
export const lmntThemePrimaryVariant = '#FB2F27';
export const lmntThemeSecondary = '#12594D';
export const lmntThemeSecondaryVariant = '#12594D';

export const primaryMap = {
    'primary-50': '#C2F4EA',
    'primary-100': '#8FD9C9',
    'primary-200': '#60C6B0',
    'primary-300': '#29AF93',
    'primary-400': '#049476',
    'primary-500': '#01866A',
    'primary-600': '#017C62',
    'primary-700': '#006A54',
    'primary-800': '#015443',
    'primary-900': '#003A2E'
};

export const secondaryMap = {
    'secondary-50': '#D2E8E4',
    'secondary-100': '#AED0CA',
    'secondary-200': '#81BCB2',
    'secondary-300': '#5EB1A3',
    'secondary-400': '#399E8D',
    'secondary-500': '#238877',
    'secondary-600': '#146D5E',
    'secondary-700': '#12594D',
    'secondary-800': '#04493E',
    'secondary-900': '#001D18'
};

export const ancillaryMap = {
    'ancillary-100': '#FB2F27'
};
