// @element/themes/lib/tokens/bayer.theme.js
export const primaryMap = {
    'primary-50': '#ffdce4',
    'primary-100': '#ffbece',
    'primary-200': '#ff7596',
    'primary-300': '#ff3162',
    'primary-400': '#de0043',
    'primary-500': '#bf124b',
    'primary-600': '#a02453',
    'primary-700': '#81365b',
    'primary-800': '#624963',
    'primary-900': '#443247'
};

export const secondaryMap = {
    'secondary-50': '#e7f8ff',
    'secondary-100': '#cdf1ff',
    'secondary-200': '#a4e7ff',
    'secondary-300': '#3eccff',
    'secondary-400': '#00bcff',
    'secondary-500': '#0091df',
    'secondary-600': '#007cb8',
    'secondary-700': '#006f9b',
    'secondary-800': '#00617f',
    'secondary-900': '#10384f'
};

export const dangerMap = {
    'danger-50': '#ffe4cd',
    'danger-100': '#ffc18d',
    'danger-200': '#fea55b',
    'danger-300': '#fd8d30',
    'danger-400': '#f9780d',
    'danger-500': '#e56f0f',
    'danger-600': '#ce630b',
    'danger-700': '#b85302',
    'danger-800': '#974401',
    'danger-900': '#652d01'
};

export const successMap = {
    'success-50': '#eeffef',
    'success-100': '#d3ffd7',
    'success-200': '#a9f6af',
    'success-300': '#6fef7a',
    'success-400': '#29e539',
    'success-500': '#13c545',
    'success-600': '#0ea01a',
    'success-700': '#01830c',
    'success-800': '#075a0e',
    'success-900': '#053709'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#1e5474',
    'dataVizComparative-100': '#00bcff',
    'dataVizComparative-200': '#0d683b',
    'dataVizComparative-300': '#89d329',
    'dataVizComparative-400': '#6a4c72',
    'dataVizComparative-500': '#cba0d7',
    'dataVizComparative-600': '#980733',
    'dataVizComparative-700': '#de0043',
    'dataVizComparative-800': '#eaeaea',
    'dataVizComparative-900': '#5c5c5c'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#092b3e',
    'dataVizSequential1-100': '#10384f',
    'dataVizSequential1-200': '#00617f',
    'dataVizSequential1-300': '#0075a6',
    'dataVizSequential1-400': '#0091df',
    'dataVizSequential1-500': '#00bcff',
    'dataVizSequential1-600': '#4fd1ff',
    'dataVizSequential1-700': '#a2e6ff',
    'dataVizSequential1-800': '#c7f0ff',
    'dataVizSequential1-900': '#e8f9ff'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#2f1533',
    'dataVizSequential2-100': '#443247',
    'dataVizSequential2-200': '#5a3663',
    'dataVizSequential2-300': '#6a4c72',
    'dataVizSequential2-400': '#8c6397',
    'dataVizSequential2-500': '#b087bb',
    'dataVizSequential2-600': '#cba0d7',
    'dataVizSequential2-700': '#e5bfef',
    'dataVizSequential2-800': '#f7dcfe',
    'dataVizSequential2-900': '#fbebff'
};

export const ancillaryMap = {
    'ancillary-100': '#89d329',
    'ancillary-200': '#00bcff',
    'ancillary-300': '#624963',
    'ancillary-400': '#ff3162'
};
