// @element/themes/lib/tokens/bayerplusrewards.theme.js
export const primaryMap = {
    'primary-50': '#e9f7d8',
    'primary-100': '#bee78a',
    'primary-200': '#89d329',
    'primary-300': '#7dc921',
    'primary-400': '#66b512',
    'primary-500': '#59a31a',
    'primary-600': '#3f802a',
    'primary-700': '#2b6636',
    'primary-800': '#13532b',
    'primary-900': '#004422'
};

export const secondaryMap = {
    'secondary-50': '#e7f8ff',
    'secondary-100': '#cdf1ff',
    'secondary-200': '#a4e7ff',
    'secondary-300': '#3eccff',
    'secondary-400': '#00bcff',
    'secondary-500': '#009be6',
    'secondary-600': '#007cb8',
    'secondary-700': '#006f9b',
    'secondary-800': '#00617f',
    'secondary-900': '#10384f'
};

export const dangerMap = {
    'danger-50': '#ffe3ca',
    'danger-100': '#ffbf84',
    'danger-200': '#ffa148',
    'danger-300': '#ff8700',
    'danger-400': '#ff6f00',
    'danger-500': '#f56700',
    'danger-600': '#dd5c00',
    'danger-700': '#c64c00',
    'danger-800': '#a23e00',
    'danger-900': '#6d2900'
};

export const successMap = {
    'success-50': '#d0ffe6',
    'success-100': '#b1ffd5',
    'success-200': '#95fec5',
    'success-300': '#82ffbb',
    'success-400': '#00ff75',
    'success-500': '#00e56a',
    'success-600': '#00d964',
    'success-700': '#01c059',
    'success-800': '#009d48',
    'success-900': '#007536'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#1e5474',
    'dataVizComparative-100': '#00bcff',
    'dataVizComparative-200': '#0d683b',
    'dataVizComparative-300': '#89d329',
    'dataVizComparative-400': '#6a4c72',
    'dataVizComparative-500': '#cba0d7',
    'dataVizComparative-600': '#980733',
    'dataVizComparative-700': '#de0043',
    'dataVizComparative-800': '#eaeaea',
    'dataVizComparative-900': '#5c5c5c'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#092b3e',
    'dataVizSequential1-100': '#10384f',
    'dataVizSequential1-200': '#00617f',
    'dataVizSequential1-300': '#0075a6',
    'dataVizSequential1-400': '#0091df',
    'dataVizSequential1-500': '#00bcff',
    'dataVizSequential1-600': '#4fd1ff',
    'dataVizSequential1-700': '#a2e6ff',
    'dataVizSequential1-800': '#c7f0ff',
    'dataVizSequential1-900': '#e8f9ff'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#012b16',
    'dataVizSequential2-100': '#004422',
    'dataVizSequential2-200': '#2b6636',
    'dataVizSequential2-300': '#108000',
    'dataVizSequential2-400': '#66b512',
    'dataVizSequential2-500': '#89d329',
    'dataVizSequential2-600': '#a2df52',
    'dataVizSequential2-700': '#cbf594',
    'dataVizSequential2-800': '#d8f8ab',
    'dataVizSequential2-900': '#e9f9d4'
};

export const ancillaryMap = {
    'ancillary-100': '#d30f4b',
    'ancillary-200': '#624963',
    'ancillary-300': '#443247'
};
