// @element/themes/lib/tokens/channel.theme.js
export const primaryMap = {
    'primary-50': '#c3d9d8',
    'primary-100': '#85b6b4',
    'primary-200': '#5ea09d',
    'primary-300': '#429390',
    'primary-400': '#27807c',
    'primary-500': '#006c67',
    'primary-600': '#01625d',
    'primary-700': '#015753',
    'primary-800': '#014d49',
    'primary-900': '#003c37'
};

export const secondaryMap = {
    'secondary-50': '#f8f9e5',
    'secondary-100': '#edf0bd',
    'secondary-200': '#e1e691',
    'secondary-300': '#d0d950',
    'secondary-400': '#c1cd23',
    'secondary-500': '#bbc613',
    'secondary-600': '#afbb0b',
    'secondary-700': '#a6b01e',
    'secondary-800': '#9ba516',
    'secondary-900': '#8f990d'
};

export const ancillaryMap = {
    'ancillary-100': '#e8941a',
    'ancillary-200': '#54301a'
};

export const successMap = {
    'success-50': '#EEFFEF',
    'success-100': '#D3FFD7',
    'success-200': '#A9F6AF',
    'success-300': '#6FEF7A',
    'success-400': '#29E539',
    'success-500': '#13C545',
    'success-600': '#0EA01A',
    'success-700': '#01830C',
    'success-800': '#075A0E',
    'success-900': '#053709'
};

export const dangerMap = {
    'danger-50': '#FFEAED',
    'danger-100': '#ffccce',
    'danger-200': '#ee9894',
    'danger-300': '#e3716a',
    'danger-400': '#ec5145',
    'danger-500': '#ef4227',
    'danger-600': '#e13727',
    'danger-700': '#cf2d22',
    'danger-800': '#c2261b',
    'danger-900': '#b3190d'
};

export const ddataVizComparativeMap = {
    'dataVizComparative-50': '#006C67',
    'dataVizComparative-100': '#C2CD23',
    'dataVizComparative-200': '#3E9E99',
    'dataVizComparative-300': '#D9E53C',
    'dataVizComparative-400': '#36BEB8',
    'dataVizComparative-500': '#DDF17F',
    'dataVizComparative-600': '#66E0DB',
    'dataVizComparative-700': '#6A706F',
    'dataVizComparative-800': '#E3E5E7',
    'dataVizComparative-900': '#424444'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#006C67',
    'dataVizSequential1-100': '#177974',
    'dataVizSequential1-200': '#2E8682',
    'dataVizSequential1-300': '#459490',
    'dataVizSequential1-400': '#5CA19E',
    'dataVizSequential1-500': '#73AEAC',
    'dataVizSequential1-600': '#8BBCB9',
    'dataVizSequential1-700': '#A2C9C7',
    'dataVizSequential1-800': '#BADDDB',
    'dataVizSequential1-900': '#D0EEEC'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#A2AD09',
    'dataVizSequential2-100': '#AFBB0B',
    'dataVizSequential2-200': '#C1CD23',
    'dataVizSequential2-300': '#CCD73E',
    'dataVizSequential2-400': '#D9E34F',
    'dataVizSequential2-500': '#E3ED63',
    'dataVizSequential2-600': '#EEF86F',
    'dataVizSequential2-700': '#F4FB9F',
    'dataVizSequential2-800': '#FAFFBF',
    'dataVizSequential2-900': '#FBFFD1'
};
