// @element/themes/lib/tokens/dad.theme.js
export const primaryMap = {
    'primary-50': '#fcfbf9',
    'primary-100': '#f3f2ed',
    'primary-200': '#e9e8e3',
    'primary-300': '#d5d2c9',
    'primary-400': '#b3afa9',
    'primary-500': '#8f8e8a',
    'primary-600': '#78756e',
    'primary-700': '#64615b',
    'primary-800': '#45423c',
    'primary-900': '#24211c'
};

export const secondaryMap = {
    'secondary-50': '#fdfdfb',
    'secondary-100': '#fafbf3',
    'secondary-200': '#f4f5ed',
    'secondary-300': '#eceee6',
    'secondary-400': '#dedfd7',
    'secondary-500': '#bbbcb4',
    'secondary-600': '#9c9d95',
    'secondary-700': '#73746d',
    'secondary-800': '#5f6059',
    'secondary-900': '#40413b'
};

export const ancillaryMap = {
    'ancillary-100': '#ffca05',
    'ancillary-200': '#8e3708',
    'ancillary-300': '#02674d'
};

export const successMap = {
    'success-50': '#EEFFEF',
    'success-100': '#D3FFD7',
    'success-200': '#A9F6AF',
    'success-300': '#6FEF7A',
    'success-400': '#29E539',
    'success-500': '#13C545',
    'success-600': '#0EA01A',
    'success-700': '#01830C',
    'success-800': '#075A0E',
    'success-900': '#053709'
};

export const dangerMap = {
    'danger-50': '#ffeaed',
    'danger-100': '#ffccce',
    'danger-200': '#ee9894',
    'danger-300': '#e3716a',
    'danger-400': '#ec5145',
    'danger-500': '#ef4227',
    'danger-600': '#e13727',
    'danger-700': '#cf2d22',
    'danger-800': '#c2261b',
    'danger-900': '#b3190d'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#1E180D',
    'dataVizComparative-100': '#9E9077',
    'dataVizComparative-200': '#52452E',
    'dataVizComparative-300': '#D1C8B7',
    'dataVizComparative-400': '#616161',
    'dataVizComparative-500': '#D47001',
    'dataVizComparative-600': '#803500',
    'dataVizComparative-700': '#D1C8B7',
    'dataVizComparative-800': '#939393',
    'dataVizComparative-900': '#EAEAEA'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#1E180D',
    'dataVizSequential1-100': '#372E1B',
    'dataVizSequential1-200': '#52452E',
    'dataVizSequential1-300': '#6C5C43',
    'dataVizSequential1-400': '#86765B',
    'dataVizSequential1-500': '#9E9077',
    'dataVizSequential1-600': '#C3B59D',
    'dataVizSequential1-700': '#EAE1D2',
    'dataVizSequential1-800': '#F7EDDE',
    'dataVizSequential1-900': '#FFF7EA'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#372E1B',
    'dataVizSequential2-100': '#49412F',
    'dataVizSequential2-200': '#5B5444',
    'dataVizSequential2-300': '#6D6759',
    'dataVizSequential2-400': '#7F7A6D',
    'dataVizSequential2-500': '#918D82',
    'dataVizSequential2-600': '#A4A097',
    'dataVizSequential2-700': '#B6B3AC',
    'dataVizSequential2-800': '#CBC8C0',
    'dataVizSequential2-900': '#DFDCD3'
};
