// @element/themes/lib/tokens/dataenablement.theme.js
export const primaryMap = {
    'primary-50': '#B8F8DE',
    'primary-100': '#84E2BC',
    'primary-200': '#68CAA3',
    'primary-300': '#55B68F',
    'primary-400': '#419D78',
    'primary-500': '#368163',
    'primary-600': '#25674D',
    'primary-700': '#18573E',
    'primary-800': '#0F4630',
    'primary-900': '#032316'
};

export const secondaryMap = {
    'secondary-50': '#D6E9FF',
    'secondary-100': '#9FCDFF',
    'secondary-200': '#72B5FF',
    'secondary-300': '#5BA5F7',
    'secondary-400': '#4190E9',
    'secondary-500': '#2F7DD5',
    'secondary-600': '#1D6BC1',
    'secondary-700': '#255C99',
    'secondary-800': '#14506A',
    'secondary-900': '#0B3C49'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#419D78',
    'dataVizComparative-100': '#0B3C49',
    'dataVizComparative-200': '#84E2BC',
    'dataVizComparative-300': '#255C99',
    'dataVizComparative-400': '#55B68F',
    'dataVizComparative-500': '#9FCCFF',
    'dataVizComparative-600': '#368163',
    'dataVizComparative-700': '#72B5FF',
    'dataVizComparative-800': '#18573E',
    'dataVizComparative-900': '#2F7DD5'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#032316',
    'dataVizSequential1-100': '#0F4630',
    'dataVizSequential1-200': '#18573E',
    'dataVizSequential1-300': '#25674D',
    'dataVizSequential1-400': '#368163',
    'dataVizSequential1-500': '#419D78',
    'dataVizSequential1-600': '#55B68F',
    'dataVizSequential1-700': '#68CAA2',
    'dataVizSequential1-800': '#84E2BC',
    'dataVizSequential1-900': '#B8F8DE'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#0B3C49',
    'dataVizSequential2-100': '#14506A',
    'dataVizSequential2-200': '#255C99',
    'dataVizSequential2-300': '#1D6BC1',
    'dataVizSequential2-400': '#2F7DD5',
    'dataVizSequential2-500': '#4090E9',
    'dataVizSequential2-600': '#5BA5F7',
    'dataVizSequential2-700': '#72B5FF',
    'dataVizSequential2-800': '#9FCCFF',
    'dataVizSequential2-900': '#D6E9FF'
};

export const dangerMap = {
    'danger-50': '#FFEAED',
    'danger-100': '#FFCCCE',
    'danger-200': '#EE9894',
    'danger-300': '#E3716A',
    'danger-400': '#EC5145',
    'danger-500': '#EF4227',
    'danger-600': '#E13727',
    'danger-700': '#CF2D22',
    'danger-800': '#C2261B',
    'danger-900': '#B3190D'
};

export const successMap = {
    'success-50': '#eeffef',
    'success-100': '#d3ffd7',
    'success-200': '#a9f6af',
    'success-300': '#6fef7a',
    'success-400': '#29e539',
    'success-500': '#13c545',
    'success-600': '#0ea01a',
    'success-700': '#01830c',
    'success-800': '#075a0e',
    'success-900': '#053709'
};

export const ancillaryMap = {
    'ancillary-100': '#CE5F27',
    'ancillary-200': '#F2C14E'
};
