// @element/themes/lib/tokens/dekalb.theme.js
export const primaryMap = {
    'primary-50': '#efeee6',
    'primary-100': '#d7d2c5',
    'primary-200': '#bcb4a2',
    'primary-300': '#9f967e',
    'primary-400': '#897e61',
    'primary-500': '#736845',
    'primary-600': '#695e3e',
    'primary-700': '#5b5034',
    'primary-800': '#4f432b',
    'primary-900': '#403420'
};

export const secondaryMap = {
    'secondary-100': '#ffefb4',
    'secondary-200': '#ffe582',
    'secondary-300': '#ffda50',
    'secondary-400': '#ffd22b',
    'secondary-500': '#ffca05',
    'secondary-600': '#ffc504',
    'secondary-700': '#ffbd04',
    'secondary-800': '#ffb703',
    'secondary-900': '#ffab01'
};

export const dangerMap = {
    'danger-50': '#ffeaed',
    'danger-100': '#ffccce',
    'danger-200': '#ee9894',
    'danger-300': '#e3716a',
    'danger-400': '#ec5145',
    'danger-500': '#ef4227',
    'danger-600': '#e13727',
    'danger-700': '#cf2d22',
    'danger-800': '#c2261b',
    'danger-900': '#b3190d'
};

export const successMap = {
    'success-50': '#EEFFEF',
    'success-100': '#D3FFD7',
    'success-200': '#A9F6AF',
    'success-300': '#6FEF7A',
    'success-400': '#29E539',
    'success-500': '#13C545',
    'success-600': '#0EA01A',
    'success-700': '#01830C',
    'success-800': '#075A0E',
    'success-900': '#053709'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#403420',
    'dataVizComparative-100': '#F3A300',
    'dataVizComparative-200': '#BCB4A2',
    'dataVizComparative-300': '#FFBD04',
    'dataVizComparative-400': '#EFEEE6',
    'dataVizComparative-500': '#FFE582',
    'dataVizComparative-600': '#4F432B',
    'dataVizComparative-700': '#FFAB01',
    'dataVizComparative-800': '#EAEAEA',
    'dataVizComparative-900': '#5C5C5C'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#403420',
    'dataVizSequential1-100': '#4F432B',
    'dataVizSequential1-200': '#5B5034',
    'dataVizSequential1-300': '#695E3E',
    'dataVizSequential1-400': '#695E3E',
    'dataVizSequential1-500': '#736845',
    'dataVizSequential1-600': '#9F967E',
    'dataVizSequential1-700': '#BCB4A2',
    'dataVizSequential1-800': '#D7D2C5',
    'dataVizSequential1-900': '#EFEEE6'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#F3A300',
    'dataVizSequential2-100': '#FFAB01',
    'dataVizSequential2-200': '#FFB703',
    'dataVizSequential2-300': '#FFBD04',
    'dataVizSequential2-400': '#FFCA05',
    'dataVizSequential2-500': '#FFCA05',
    'dataVizSequential2-600': '#FFD22B',
    'dataVizSequential2-700': '#FFE582',
    'dataVizSequential2-800': '#FFEFB4',
    'dataVizSequential2-900': '#FFF9E1'
};
