// @element/themes/lib/tokens/deltapine.theme.js
export const primaryMap = {
    'primary-50': '#e8f6ed',
    'primary-100': '#c8ead2',
    'primary-200': '#a6dcb5',
    'primary-300': '#81cf98',
    'primary-400': '#65c482',
    'primary-500': '#49b96d',
    'primary-600': '#41aa62',
    'primary-700': '#389756',
    'primary-800': '#30864b',
    'primary-900': '#226637'
};

export const secondaryMap = {
    'secondary-50': '#fff9e1',
    'secondary-100': '#fff9a8',
    'secondary-200': '#fff78a',
    'secondary-300': '#faf179',
    'secondary-400': '#f5ef42',
    'secondary-500': '#f8e025',
    'secondary-600': '#fbde04',
    'secondary-700': '#fad400',
    'secondary-800': '#f5c000',
    'secondary-900': '#ebaf36'
};

export const dangerMap = {
    'danger-50': '#ffeaed',
    'danger-100': '#ffccce',
    'danger-200': '#ee9894',
    'danger-300': '#e3716a',
    'danger-400': '#ec5145',
    'danger-500': '#ef4227',
    'danger-600': '#e13727',
    'danger-700': '#cf2d22',
    'danger-800': '#c2261b',
    'danger-900': '#b3190d'
};

export const successMap = {
    'success-50': '#EEFFEF',
    'success-100': '#D3FFD7',
    'success-200': '#A9F6AF',
    'success-300': '#6FEF7A',
    'success-400': '#29E539',
    'success-500': '#13C545',
    'success-600': '#0EA01A',
    'success-700': '#01830C',
    'success-800': '#075A0E',
    'success-900': '#053709'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#226637',
    'dataVizComparative-100': '#FFF9A8',
    'dataVizComparative-200': '#2B7E45',
    'dataVizComparative-300': '#F5EF42',
    'dataVizComparative-400': '#41AA62',
    'dataVizComparative-500': '#F5C000',
    'dataVizComparative-600': '#65C482',
    'dataVizComparative-700': '#865D0A',
    'dataVizComparative-800': '#A6DCB5',
    'dataVizComparative-900': '#939393'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#154624',
    'dataVizSequential1-100': '#226637',
    'dataVizSequential1-200': '#2B7E45',
    'dataVizSequential1-300': '#389756',
    'dataVizSequential1-400': '#41AA62',
    'dataVizSequential1-500': '#65C482',
    'dataVizSequential1-600': '#81CF98',
    'dataVizSequential1-700': '#A6DCB5',
    'dataVizSequential1-800': '#C8EAD2',
    'dataVizSequential1-900': '#E8F6ED'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#C99100',
    'dataVizSequential2-100': '#D99C00',
    'dataVizSequential2-200': '#F3AF01',
    'dataVizSequential2-300': '#F0BC00',
    'dataVizSequential2-400': '#F9C407',
    'dataVizSequential2-500': '#FFCA0B',
    'dataVizSequential2-600': '#FFDA52',
    'dataVizSequential2-700': '#FFE457',
    'dataVizSequential2-800': '#FFEC8B',
    'dataVizSequential2-900': '#FFF3B2'
};
