// @element/themes/lib/tokens/diane35.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    dangerMap,
    successMap,
    lmntThemeDanger,
    lmntThemeDangerVariant,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#D40F7D';
export const lmntThemePrimaryVariant = '#D04135';
export const lmntThemeSecondary = '#D40F7D';
export const lmntThemeSecondaryVariant = '#D40F7D';

export const primaryMap = {
    'primary-50': '#FF85C9',
    'primary-100': '#F874BE',
    'primary-200': '#F263B3',
    'primary-300': '#EC52A8',
    'primary-400': '#E6419D',
    'primary-500': '#E03092',
    'primary-600': '#DA1F87',
    'primary-700': '#D40F7D',
    'primary-800': '#B20064',
    'primary-900': '#8F0050'
};

export const secondaryMap = {
    'secondary-50': '#FF85C9',
    'secondary-100': '#F874BE',
    'secondary-200': '#F263B3',
    'secondary-300': '#EC52A8',
    'secondary-400': '#E6419D',
    'secondary-500': '#E03092',
    'secondary-600': '#DA1F87',
    'secondary-700': '#D40F7D',
    'secondary-800': '#B20064',
    'secondary-900': '#8F0050'
};

export const ancillaryMap = {
    'ancillary-100': '#D04135'
};
