// @element/themes/lib/tokens/eylea.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#003B5C';
export const lmntThemePrimaryVariant = '#FFD200';
export const lmntThemeSecondary = '#003B5C';
export const lmntThemeSecondaryVariant = '#003B5C';
export const lmntThemeDanger = '#CF2D22';
export const lmntThemeDangerVariant = '#A8150B';

export const primaryMap = {
    'primary-50': '#69A7C9',
    'primary-100': '#5A97B9',
    'primary-200': '#4B88A9',
    'primary-300': '#3C789A',
    'primary-400': '#2D698A',
    'primary-500': '#1E597B',
    'primary-600': '#0F4A6B',
    'primary-700': '#003B5C',
    'primary-800': '#002C45',
    'primary-900': '#002031'
};

export const secondaryMap = {
    'secondary-50': '#69A7C9',
    'secondary-100': '#5A97B9',
    'secondary-200': '#4B88A9',
    'secondary-300': '#3C789A',
    'secondary-400': '#2D698A',
    'secondary-500': '#1E597B',
    'secondary-600': '#0F4A6B',
    'secondary-700': '#003B5C',
    'secondary-800': '#002C45',
    'secondary-900': '#002031'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#FFD200'
};
