// @element/themes/lib/tokens/fontanelle.theme.js
export const primaryMap = {
    'primary-50': '#f2f8ff',
    'primary-100': '#d0dce9',
    'primary-200': '#8ba7c8',
    'primary-300': '#5c84b2',
    'primary-400': '#2e619c',
    'primary-500': '#003f87',
    'primary-600': '#00346f',
    'primary-700': '#002956',
    'primary-800': '#001732',
    'primary-900': '#000c19'
};

export const secondaryMap = {
    'secondary-50': '#fff8dd',
    'secondary-100': '#fff4c6',
    'secondary-200': '#fff0ae',
    'secondary-300': '#ffe77f',
    'secondary-400': '#ffde4f',
    'secondary-500': '#fcd116',
    'secondary-600': '#fcbd16',
    'secondary-700': '#eda317',
    'secondary-800': '#bc6e00',
    'secondary-900': '#592300'
};

export const dangerMap = {
    'danger-50': '#ffe4cd',
    'danger-100': '#ffc18d',
    'danger-200': '#fea55b',
    'danger-300': '#fd8d30',
    'danger-400': '#f9780d',
    'danger-500': '#e56f0f',
    'danger-600': '#ce630b',
    'danger-700': '#b85302',
    'danger-800': '#974401',
    'danger-900': '#652d01'
};

export const successMap = {
    'success-50': '#eeffef',
    'success-100': '#d3ffd7',
    'success-200': '#a9f6af',
    'success-300': '#6fef7a',
    'success-400': '#29e539',
    'success-500': '#13c545',
    'success-600': '#0ea01a',
    'success-700': '#01830c',
    'success-800': '#075a0e',
    'success-900': '#053709'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#001732',
    'dataVizComparative-100': '#FFF0AE',
    'dataVizComparative-200': '#003F87',
    'dataVizComparative-300': '#FCD116',
    'dataVizComparative-400': '#5C84B2',
    'dataVizComparative-500': '#EDA317',
    'dataVizComparative-600': '#8BA7C8',
    'dataVizComparative-700': '#BC6E00',
    'dataVizComparative-800': '#D0DCE9',
    'dataVizComparative-900': '#998D8D'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#000000',
    'dataVizSequential1-100': '#001732',
    'dataVizSequential1-200': '#002956',
    'dataVizSequential1-300': '#00346F',
    'dataVizSequential1-400': '#003F87',
    'dataVizSequential1-500': '#2E619C',
    'dataVizSequential1-600': '#5C84B2',
    'dataVizSequential1-700': '#8BA7C8',
    'dataVizSequential1-800': '#D0DCE9',
    'dataVizSequential1-900': '#F2F8FF'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#9C5B00',
    'dataVizSequential2-100': '#BC6E00',
    'dataVizSequential2-200': '#EDA317',
    'dataVizSequential2-300': '#FCBD16',
    'dataVizSequential2-400': '#FCD116',
    'dataVizSequential2-500': '#FFDE4F',
    'dataVizSequential2-600': '#FFE77F',
    'dataVizSequential2-700': '#FFF0AE',
    'dataVizSequential2-800': '#FFF4C6',
    'dataVizSequential2-900': '#FFF8DD'
};
