// @element/themes/lib/tokens/hcp.theme.js
export const lmntThemeSurface = '#fff';
export const lmntThemeBackground = lmntThemeSurface;
export const lmntThemePrimary = '#0050c7';
export const lmntThemePrimaryVariant = '#aa4eb1';
export const lmntThemeSecondary = '#10384f';
export const lmntThemeSecondaryVariant = '#10384f';
export const lmntThemeDanger = '#b85302';
export const lmntThemeDangerVariant = '#f9780d';
export const lmntThemeSuccess = '#0ea01a';
export const lmntThemeSuccessVariant = '#13c545';

export const breakpointsMap = {
    'mobile-min': '0',
    'mobile-max': '599px',
    'tablet-sm-min': '600px',
    'tablet-sm-max': '904px',
    'tablet-lg-min': '905px',
    'tablet-lg-max': '1159px',
    'desktop-sm-min': '1160px',
    'desktop-sm-max': '1495px',
    'desktop-lg-min': '1495px',
    'desktop-lg-max': '1495px',
    'desktop-xl-min': '1495px',
    'content-max-width-md': '7688px',
    'content-max-width-lg': '1070px',
    'content-max-width-xl': '1200px'
};

export const primaryMap = {
    'primary-50': '#9fbeed',
    'primary-100': '#7fa8e5',
    'primary-200': '#5f92dd',
    'primary-300': '#3f7cd6',
    'primary-400': '#1f66ce',
    'primary-500': '#0050c7',
    'primary-600': '#0048b3',
    'primary-700': '#00409f',
    'primary-800': '#00388b',
    'primary-900': '#003078'
};

export const secondaryMap = {
    'secondary-50': '#edf8ff',
    'secondary-100': '#caecff',
    'secondary-200': '#b9e5ff',
    'secondary-300': '#96d9ff',
    'secondary-400': '#74ccff',
    'secondary-500': '#40baff',
    'secondary-600': '#3599d1',
    'secondary-700': '#2f88ba',
    'secondary-800': '#23668c',
    'secondary-900': '#18445d'
};

export const dangerMap = {
    'danger-50': '#ffe4cd',
    'danger-100': '#ffc18d',
    'danger-200': '#fea55b',
    'danger-300': '#fd8d30',
    'danger-400': '#f9780d',
    'danger-500': '#e56f0f',
    'danger-600': '#ce630b',
    'danger-700': '#b85302',
    'danger-800': '#974401',
    'danger-900': '#652d01'
};

export const ancillaryMap = {
    'ancillary-100': '#89d329',
    'ancillary-200': '#00bcff',
    'ancillary-300': '#624963'
};

export const successMap = {
    'success-50': '#eeffef',
    'success-100': '#d3ffd7',
    'success-200': '#a9f6af',
    'success-300': '#6fef7a',
    'success-400': '#29e539',
    'success-500': '#13c545',
    'success-600': '#0ea01a',
    'success-700': '#01830c',
    'success-800': '#075a0e',
    'success-900': '#053709'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#1e5474',
    'dataVizComparative-100': '#00bcff',
    'dataVizComparative-200': '#0d683b',
    'dataVizComparative-300': '#89d329',
    'dataVizComparative-400': '#6a4c72',
    'dataVizComparative-500': '#cba0d7',
    'dataVizComparative-600': '#980733',
    'dataVizComparative-700': '#de0043',
    'dataVizComparative-800': '#eaeaea',
    'dataVizComparative-900': '#5c5c5c'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#092b3e',
    'dataVizSequential1-100': '#10384f',
    'dataVizSequential1-200': '#00617f',
    'dataVizSequential1-300': '#0075a6',
    'dataVizSequential1-400': '#0091df',
    'dataVizSequential1-500': '#00bcff',
    'dataVizSequential1-600': '#4fd1ff',
    'dataVizSequential1-700': '#a2e6ff',
    'dataVizSequential1-800': '#c7f0ff',
    'dataVizSequential1-900': '#e8f9ff'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#2f1533',
    'dataVizSequential2-100': '#443247',
    'dataVizSequential2-200': '#5a3663',
    'dataVizSequential2-300': '#6a4c72',
    'dataVizSequential2-400': '#8c6397',
    'dataVizSequential2-500': '#b087bb',
    'dataVizSequential2-600': '#cba0d7',
    'dataVizSequential2-700': '#e5bfef',
    'dataVizSequential2-800': '#f7dcfe',
    'dataVizSequential2-900': '#fbebff'
};

export const grayMap = {
    lightGray: '#f3f4f4',
    midGray: '#c1c1c1',
    midDarkGray: '#a1a3a3',
    darkGray: '#696969'
};
