// @element/themes/lib/tokens/jivi.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#157CBC';
export const lmntThemePrimaryVariant = ' #F0E635';
export const lmntThemeSecondary = '#2C326E';
export const lmntThemeSecondaryVariant = '#2C326E';
export const lmntThemeDanger = '#cf2d22';
export const lmntThemeDangerVariant = '#a8150b';

export const primaryMap = {
    'primary-50': '#80CEFF',
    'primary-100': '#70C2F5',
    'primary-200': '#61B6EB',
    'primary-300': '#52AAE2',
    'primary-400': '#429FD8',
    'primary-500': '#3393CF',
    'primary-600': '#2487C5',
    'primary-700': '#157CBC',
    'primary-800': '#0A69A4',
    'primary-900': '#00568C'
};

export const secondaryMap = {
    'secondary-50': '#9CA2DA',
    'secondary-100': '#8C92CA',
    'secondary-200': '#7C82BB',
    'secondary-300': '#6C72AB',
    'secondary-400': '#5C629C',
    'secondary-500': '#4C528C',
    'secondary-600': '#3C427D',
    'secondary-700': '#2C326E',
    'secondary-800': '#161C59',
    'secondary-900': '#010745'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#F0E635'
};
