// @element/themes/lib/tokens/jungseedgenetics.theme.js
export const primaryMap = {
    'primary-50': '#d6f2d7',
    'primary-100': '#a4eeab',
    'primary-200': '#73eb85',
    'primary-300': '#2ce251',
    'primary-400': '#26d653',
    'primary-500': '#2bc55b',
    'primary-600': '#2e945b',
    'primary-700': '#28724f',
    'primary-800': '#135030',
    'primary-900': '#052816'
};

export const secondaryMap = {
    'secondary-50': '#ffcacb',
    'secondary-100': '#ffaeb3',
    'secondary-200': '#ff8291',
    'secondary-300': '#fc6177',
    'secondary-400': '#f22849',
    'secondary-500': '#e4002b',
    'secondary-600': '#c50e29',
    'secondary-700': '#a11b28',
    'secondary-800': '#7a2126',
    'secondary-900': '#521f1f'
};

export const dangerMap = {
    'danger-50': '#ffc26a',
    'danger-100': '#ffa528',
    'danger-200': '#fd9327',
    'danger-300': '#fc8026',
    'danger-400': '#ee6820',
    'danger-500': '#e35e1b',
    'danger-600': '#d35415',
    'danger-700': '#bf470e',
    'danger-800': '#862b00',
    'danger-900': '#642000'
};

export const successMap = {
    'success-50': '#EEFFEF',
    'success-100': '#D3FFD7',
    'success-200': '#A9F6AF',
    'success-300': '#6FEF7A',
    'success-400': '#29E539',
    'success-500': '#13C545',
    'success-600': '#0EA01A',
    'success-700': '#01830C',
    'success-800': '#075A0E',
    'success-900': '#053709'
};

export const ddataVipComparativeMap = {
    'dataVizComparative-50': '#052816',
    'dataVizComparative-100': '#F22849',
    'dataVizComparative-200': '#28724F',
    'dataVizComparative-300': '#C50E29',
    'dataVizComparative-400': '#2BC55B',
    'dataVizComparative-500': '#A11B28',
    'dataVizComparative-600': '#2CE251',
    'dataVizComparative-700': '#521F1F',
    'dataVizComparative-800': '#A4EEAB',
    'dataVizComparative-900': '#A8A8A8'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#01140A',
    'dataVizSequential1-100': '#135030',
    'dataVizSequential1-200': '#28724F',
    'dataVizSequential1-300': '#2E945B',
    'dataVizSequential1-400': '#22BF53',
    'dataVizSequential1-500': '#26D653',
    'dataVizSequential1-600': '#2CE251',
    'dataVizSequential1-700': '#73EB85',
    'dataVizSequential1-800': '#A4EEAB',
    'dataVizSequential1-900': '#D6F2D7'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#350C0C',
    'dataVizSequential2-100': '#521F1F',
    'dataVizSequential2-200': '#7A2126',
    'dataVizSequential2-300': '#A11B28',
    'dataVizSequential2-400': '#C50E29',
    'dataVizSequential2-500': '#E4002B',
    'dataVizSequential2-600': '#F22849',
    'dataVizSequential2-700': '#FC6177',
    'dataVizSequential2-800': '#FF8291',
    'dataVizSequential2-900': '#FFAEB3'
};
