// @element/themes/lib/tokens/kerendia.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    dangerMap,
    successMap,
    lmntThemeDanger,
    lmntThemeDangerVariant,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#B5002B';
export const lmntThemePrimaryVariant = ' #B5002B';
export const lmntThemeSecondary = '#B5002B';
export const lmntThemeSecondaryVariant = '#B5002B';

export const primaryMap = {
    'primary-50': '#FF83A0',
    'primary-100': '#FF517A',
    'primary-200': '#EA3862',
    'primary-300': '#D9254F',
    'primary-400': '#C60F3A',
    'primary-500': '#B5002B',
    'primary-600': '#A10127',
    'primary-700': '#920123',
    'primary-800': '#7C011E',
    'primary-900': '#68021A'
};

export const secondaryMap = {
    'secondary-50': '#FF83A0',
    'secondary-100': '#FF517A',
    'secondary-200': '#EA3862',
    'secondary-300': '#D9254F',
    'secondary-400': '#C60F3A',
    'secondary-500': '#B5002B',
    'secondary-600': '#A10127',
    'secondary-700': '#920123',
    'secondary-800': '#7C011E',
    'secondary-900': '#68021A'
};

export const ancillaryMap = {
    'ancillary-100': '#B5002B'
};
