// @element/themes/lib/tokens/kovaltry.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#0A66AF';
export const lmntThemePrimaryVariant = ' #66A242';
export const lmntThemeSecondary = '#00265D';
export const lmntThemeSecondaryVariant = '#00265D';
export const lmntThemeDanger = '#CF2D22';
export const lmntThemeDangerVariant = '#A8150B';

export const primaryMap = {
    'primary-50': '#7AC2FF',
    'primary-100': '#6AB4F3',
    'primary-200': '#5AA7E8',
    'primary-300': '#4A9ADC',
    'primary-400': '#3A8DD1',
    'primary-500': '#2A80C5',
    'primary-600': '#1A73BA',
    'primary-700': '#0A66AF',
    'primary-800': '#014D89',
    'primary-900': '#00355F'
};

export const secondaryMap = {
    'secondary-50': '#719EDE',
    'secondary-100': '#608CCB',
    'secondary-200': '#507BB9',
    'secondary-300': '#406AA6',
    'secondary-400': '#305994',
    'secondary-500': '#204881',
    'secondary-600': '#10376F',
    'secondary-700': '#00265D',
    'secondary-800': '#001B41',
    'secondary-900': '#001636'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#66A242'
};
