// @element/themes/lib/tokens/kyleena.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    dangerMap,
    successMap,
    lmntThemeDanger,
    lmntThemeDangerVariant,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#AD0065';
export const lmntThemePrimaryVariant = '#4B5195';
export const lmntThemeSecondary = '#4B5195';
export const lmntThemeSecondaryVariant = '#4B5195';

export const primaryMap = {
    'primary-50': '#FCAFDC',
    'primary-100': '#F185C4',
    'primary-200': '#E76DB4',
    'primary-300': '#DC52A3',
    'primary-400': '#D23C94',
    'primary-500': '#C72985',
    'primary-600': '#BF0873',
    'primary-700': '#AD0065',
    'primary-800': '#9A015A',
    'primary-900': '#7C0048'
};

export const secondaryMap = {
    'secondary-50': '#D0D2EE',
    'secondary-100': '#AEB1DF',
    'secondary-200': '#989CD0',
    'secondary-300': '#858AC5',
    'secondary-400': '#757AB6',
    'secondary-500': '#5E64A7',
    'secondary-600': '#4B5195',
    'secondary-700': '#373D86',
    'secondary-800': '#232A76',
    'secondary-900': '#0E134E'
};

export const ancillaryMap = {
    'ancillary-100': '#4B5195'
};
