// @element/themes/lib/tokens/mirena.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#007D86';
export const lmntThemePrimaryVariant = ' #AFBB4B';
export const lmntThemeSecondary = '#007D86';
export const lmntThemeSecondaryVariant = '#007D86';
export const lmntThemeDanger = '#CF2D22';
export const lmntThemeDangerVariant = '#A8150B';

export const primaryMap = {
    'primary-50': '#7DE4EC',
    'primary-100': '#6BD5DD',
    'primary-200': '#59C6CE',
    'primary-300': '#47B7C0',
    'primary-400': '#35A9B1',
    'primary-500': '#239AA3',
    'primary-600': '#118B94',
    'primary-700': '#007D86',
    'primary-800': '#006269',
    'primary-900': '#004348'
};

export const secondaryMap = {
    'secondary-50': '#7DE4EC',
    'secondary-100': '#6BD5DD',
    'secondary-200': '#59C6CE',
    'secondary-300': '#47B7C0',
    'secondary-400': '#35A9B1',
    'secondary-500': '#239AA3',
    'secondary-600': '#118B94',
    'secondary-700': '#007D86',
    'secondary-800': '#006269',
    'secondary-900': '#004348'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#AFBB4B'
};
