// @element/themes/lib/tokens/monsoy.theme.js
export const primaryMap = {
    'primary-50': '#fbf3ea',
    'primary-100': '#f7e7d7',
    'primary-200': '#efcfb0',
    'primary-300': '#e3ad76',
    'primary-400': '#da964f',
    'primary-500': '#cf7418',
    'primary-600': '#bf6607',
    'primary-700': '#af5900',
    'primary-800': '#a25302',
    'primary-900': '#8f4700'
};

export const secondaryMap = {
    'secondary-50': '#e7f1ee',
    'secondary-100': '#b9d5cc',
    'secondary-200': '#8bbaaa',
    'secondary-300': '#6eab96',
    'secondary-400': '#4a9f82',
    'secondary-500': '#2b8d6a',
    'secondary-600': '#0c8057',
    'secondary-700': '#08744e',
    'secondary-800': '#006944',
    'secondary-900': '#045137'
};

export const successMap = {
    'success-50': '#EEFFEF',
    'success-100': '#D3FFD7',
    'success-200': '#A9F6AF',
    'success-300': '#6FEF7A',
    'success-400': '#29E539',
    'success-500': '#13C545',
    'success-600': '#0EA01A',
    'success-700': '#01830C',
    'success-800': '#075A0E',
    'success-900': '#053709'
};

export const dangerMap = {
    'danger-50': '#FFEAED',
    'danger-100': '#ffccce',
    'danger-200': '#ee9894',
    'danger-300': '#e3716a',
    'danger-400': '#ec5145',
    'danger-500': '#ef4227',
    'danger-600': '#e13727',
    'danger-700': '#cf2d22',
    'danger-800': '#c2261b',
    'danger-900': '#b3190d'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#8F4700',
    'dataVizComparative-100': '#8BBAAA',
    'dataVizComparative-200': '#AF5900',
    'dataVizComparative-300': '#4A9F82',
    'dataVizComparative-400': '#CF7418',
    'dataVizComparative-500': '#006944',
    'dataVizComparative-600': '#E3AD76',
    'dataVizComparative-700': '#045137',
    'dataVizComparative-800': '#F7E7D7',
    'dataVizComparative-900': '#848484'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#8F4700',
    'dataVizSequential1-100': '#A25302',
    'dataVizSequential1-200': '#AF5900',
    'dataVizSequential1-300': '#BF6607',
    'dataVizSequential1-400': '#CF7418',
    'dataVizSequential1-500': '#DA964F',
    'dataVizSequential1-600': '#E3AD76',
    'dataVizSequential1-700': '#EFCFB0',
    'dataVizSequential1-800': '#F7E7D7',
    'dataVizSequential1-900': '#FBF3EA'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#045137',
    'dataVizSequential2-100': '#006944',
    'dataVizSequential2-200': '#08744E',
    'dataVizSequential2-300': '#0C8057',
    'dataVizSequential2-400': '#2B8D6A',
    'dataVizSequential2-500': '#4A9F82',
    'dataVizSequential2-600': '#6EAB96',
    'dataVizSequential2-700': '#8BBAAA',
    'dataVizSequential2-800': '#B9D5CC',
    'dataVizSequential2-900': '#E7F1EE'
};
