// @element/themes/lib/tokens/nexavar.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#3A73D4';
export const lmntThemePrimaryVariant = ' #AF8520';
export const lmntThemeSecondary = '#1E285A';
export const lmntThemeSecondaryVariant = '#1E285A';
export const lmntThemeDanger = '#CF2D22';
export const lmntThemeDangerVariant = '#A8150B';

export const primaryMap = {
    'primary-50': '#AAC8FD',
    'primary-100': '#9ABBF7',
    'primary-200': '#8AAFF1',
    'primary-300': '#7AA3EB',
    'primary-400': '#6A97E5',
    'primary-500': '#5A8BDF',
    'primary-600': '#4A7FD9',
    'primary-700': '#3A73D4',
    'primary-800': '#1D55B6',
    'primary-900': '#003898'
};

export const secondaryMap = {
    'secondary-50': '#A9B4EA',
    'secondary-100': '#95A0D5',
    'secondary-200': '#818CC0',
    'secondary-300': '#6D78AC',
    'secondary-400': '#596497',
    'secondary-500': '#455083',
    'secondary-600': '#313C6E',
    'secondary-700': '#1E285A',
    'secondary-800': '#121B47',
    'secondary-900': '#060E35'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#AF8520'
};
