// @element/themes/lib/tokens/nuqeba.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#223C78';
export const lmntThemePrimaryVariant = '#F16623';
export const lmntThemeSecondary = '#2C7DA5';
export const lmntThemeSecondaryVariant = '#2C7DA5';
export const lmntThemeDanger = '#CF2D22';
export const lmntThemeDangerVariant = '#A8150B';

export const primaryMap = {
    'primary-50': '#9EB1DE',
    'primary-100': '#8CA1CF',
    'primary-200': '#7A90C0',
    'primary-300': '#687FB2',
    'primary-400': '#576EA3',
    'primary-500': '#455D95',
    'primary-600': '#334C86',
    'primary-700': '#223C78',
    'primary-800': '#0E2763',
    'primary-900': '#00123D'
};

export const secondaryMap = {
    'secondary-50': '#9FD6F1',
    'secondary-100': '#8EC9E6',
    'secondary-200': '#7EBCDB',
    'secondary-300': '#6DAFD0',
    'secondary-400': '#5DA3C5',
    'secondary-500': '#4C96BA',
    'secondary-600': '#3C89AF',
    'secondary-700': '#2C7DA5',
    'secondary-800': '#085E88',
    'secondary-900': '#003F5F'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#F16623'
};
