// @element/themes/lib/tokens/qlaira.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#3E72B6';
export const lmntThemePrimaryVariant = ' #F1BD37';
export const lmntThemeSecondary = '#3E72B6';
export const lmntThemeSecondaryVariant = '#3E72B6';
export const lmntThemeDanger = '#CF2D22';
export const lmntThemeDangerVariant = '#A8150B';

export const primaryMap = {
    'primary-50': '#96BEF4',
    'primary-100': '#6E9FE0',
    'primary-200': '#5C8FD2',
    'primary-300': '#4B7FC3',
    'primary-400': '#3E72B6',
    'primary-500': '#3266AA',
    'primary-600': '#265DA5',
    'primary-700': '#1C5197',
    'primary-800': '#134588',
    'primary-900': '#063473'
};

export const secondaryMap = {
    'secondary-50': '#96BEF4',
    'secondary-100': '#6E9FE0',
    'secondary-200': '#5C8FD2',
    'secondary-300': '#4B7FC3',
    'secondary-400': '#3E72B6',
    'secondary-500': '#3266AA',
    'secondary-600': '#265DA5',
    'secondary-700': '#1C5197',
    'secondary-800': '#134588',
    'secondary-900': '#063473'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#F1BD37'
};
