// @element/themes/lib/tokens/rea.theme.js
export const primaryMap = {
    'primary-50': '#ebf0f5',
    'primary-100': '#d7e2eb',
    'primary-200': '#9bb7cd',
    'primary-300': '#739ab9',
    'primary-400': '#4b7da5',
    'primary-500': '#236192',
    'primary-600': '#1d5078',
    'primary-700': '#173e5d',
    'primary-800': '#102d43',
    'primary-900': '#000'
};
export const secondaryMap = {
    'secondary-50': '#f5ebed',
    'secondary-100': '#e2c4cb',
    'secondary-200': '#d09ca9',
    'secondary-300': '#bd7587',
    'secondary-400': '#aa4e65',
    'secondary-500': '#982743',
    'secondary-600': '#7d2037',
    'secondary-700': '#6f1d31',
    'secondary-800': '#531625',
    'secondary-900': '#2a0b13'
};

export const successMap = {
    'success-50': '#EEFFEF',
    'success-100': '#D3FFD7',
    'success-200': '#A9F6AF',
    'success-300': '#6FEF7A',
    'success-400': '#29E539',
    'success-500': '#13C545',
    'success-600': '#0EA01A',
    'success-700': '#01830C',
    'success-800': '#075A0E',
    'success-900': '#053709'
};

export const dangerMap = {
    'danger-50': '#FFEAED',
    'danger-100': '#ffccce',
    'danger-200': '#ee9894',
    'danger-300': '#e3716a',
    'danger-400': '#ec5145',
    'danger-500': '#ef4227',
    'danger-600': '#e13727',
    'danger-700': '#cf2d22',
    'danger-800': '#c2261b',
    'danger-900': '#b3190d'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#173E5D',
    'dataVizComparative-100': '#AA4E65',
    'dataVizComparative-200': '#1D5078',
    'dataVizComparative-300': '#982743',
    'dataVizComparative-400': '#4B7DA5',
    'dataVizComparative-500': '#7D2037',
    'dataVizComparative-600': '#D7E2EB',
    'dataVizComparative-700': '#531625',
    'dataVizComparative-800': '#000000',
    'dataVizComparative-900': '#8F8F8F'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#000000',
    'dataVizSequential1-100': '#102D43',
    'dataVizSequential1-200': '#173E5D',
    'dataVizSequential1-300': '#1D5078',
    'dataVizSequential1-400': '#236192',
    'dataVizSequential1-500': '#4B7DA5',
    'dataVizSequential1-600': '#739AB9',
    'dataVizSequential1-700': '#9BB7CD',
    'dataVizSequential1-800': '#D7E2EB',
    'dataVizSequential1-900': '#EBF0F5'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#2A0B13',
    'dataVizSequential2-100': '#531625',
    'dataVizSequential2-200': '#6F1D31',
    'dataVizSequential2-300': '#7D2037',
    'dataVizSequential2-400': '#982743',
    'dataVizSequential2-500': '#AA4E65',
    'dataVizSequential2-600': '#BD7587',
    'dataVizSequential2-700': '#D09CA9',
    'dataVizSequential2-800': '#E2C4CB',
    'dataVizSequential2-900': '#F5EBED'
};
