// @element/themes/lib/tokens/stewart.theme.js
export const primaryMap = {
    'primary-50': '#ffcacc',
    'primary-100': '#ffa5a9',
    'primary-200': '#fe8389',
    'primary-300': '#f8656d',
    'primary-400': '#e23741',
    'primary-500': '#d22630',
    'primary-600': '#b42823',
    'primary-700': '#93272c',
    'primary-800': '#702426',
    'primary-900': '#4f1a1b'
};

export const secondaryMap = {
    'secondary-50': '#fffcd2',
    'secondary-100': '#fffab1',
    'secondary-200': '#fff382',
    'secondary-300': '#ffea55',
    'secondary-400': '#ffe02b',
    'secondary-500': '#fad306',
    'secondary-600': '#e2c216',
    'secondary-700': '#c6ad23',
    'secondary-800': '#a5942c',
    'secondary-900': '#81752f'
};

export const successMap = {
    'success-50': '#EEFFEF',
    'success-100': '#D3FFD7',
    'success-200': '#A9F6AF',
    'success-300': '#6FEF7A',
    'success-400': '#29E539',
    'success-500': '#13C545',
    'success-600': '#0EA01A',
    'success-700': '#01830C',
    'success-800': '#075A0E',
    'success-900': '#053709'
};

export const dangerMap = {
    'danger-50': '#FFEAED',
    'danger-100': '#ffccce',
    'danger-200': '#ee9894',
    'danger-300': '#e3716a',
    'danger-400': '#ec5145',
    'danger-500': '#ef4227',
    'danger-600': '#e13727',
    'danger-700': '#cf2d22',
    'danger-800': '#c2261b',
    'danger-900': '#b3190d'
};

export const ancillaryMap = {
    'ancillary-100': '#3AB7A2',
    'ancillary-200': '#F6A649',
    'ancillary-300': '#F886E0'
};

export const ddataVipComparativeMap = {
    'dataVizComparative-50': '#240F67',
    'dataVizComparative-100': '#2D72C4',
    'dataVizComparative-200': '#89C9B7',
    'dataVizComparative-300': '#785E9F',
    'dataVizComparative-400': '#3B8EC8',
    'dataVizComparative-500': '#B7DBBC',
    'dataVizComparative-600': '#4153AF',
    'dataVizComparative-700': '#47A3C4',
    'dataVizComparative-800': '#DFECCB',
    'dataVizComparative-900': '#6332AD'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#261643',
    'dataVizSequential1-100': '#422575',
    'dataVizSequential1-200': '#6540AA',
    'dataVizSequential1-300': '#7E52D0',
    'dataVizSequential1-400': '#9A75DD',
    'dataVizSequential1-500': '#AF8BEF',
    'dataVizSequential1-600': '#C6A6FF',
    'dataVizSequential1-700': '#D8C1FF',
    'dataVizSequential1-800': '#E8DBFF',
    'dataVizSequential1-900': '#F5EFFF'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#052D44',
    'dataVizSequential2-100': '#08517B',
    'dataVizSequential2-200': '#1774AA',
    'dataVizSequential2-300': '#319BD8',
    'dataVizSequential2-400': '#40BAFF',
    'dataVizSequential2-500': '#75CDFF',
    'dataVizSequential2-600': '#97E0FF',
    'dataVizSequential2-700': '#A9F0FF',
    'dataVizSequential2-800': '#C3FBFF',
    'dataVizSequential2-900': '#DDFFFD'
};
