// @element/themes/lib/tokens/stivarga.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#00626B';
export const lmntThemePrimaryVariant = ' #00BCB5';
export const lmntThemeSecondary = '#00626B';
export const lmntThemeSecondaryVariant = '#00626B';
export const lmntThemeDanger = '#CF2D22';
export const lmntThemeDangerVariant = '#A8150B';

export const primaryMap = {
    'primary-50': '#8FD6DF',
    'primary-100': '#7AC5CE',
    'primary-200': '#66B4BD',
    'primary-300': '#51A4AD',
    'primary-400': '#3D939C',
    'primary-500': '#28838C',
    'primary-600': '#14727B',
    'primary-700': '#00626B',
    'primary-800': '#01555C',
    'primary-900': '#004950'
};

export const secondaryMap = {
    'secondary-50': '#8FD6DF',
    'secondary-100': '#7AC5CE',
    'secondary-200': '#66B4BD',
    'secondary-300': '#51A4AD',
    'secondary-400': '#3D939C',
    'secondary-500': '#28838C',
    'secondary-600': '#14727B',
    'secondary-700': '#00626B',
    'secondary-800': '#01555C',
    'secondary-900': '#004950'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#00BCB5'
};
