// @element/themes/lib/tokens/vegstate.theme.js
export const primaryMap = {
    'primary-50': '#e7f8ff',
    'primary-100': '#cdf1ff',
    'primary-200': '#a4e7ff',
    'primary-300': '#3eccff',
    'primary-400': '#00beff',
    'primary-500': '#009be6',
    'primary-600': '#007cb8',
    'primary-700': '#007cb8',
    'primary-800': '#00617f',
    'primary-900': '#10384f'
};

export const secondaryMap = {
    'secondary-50': '#e9f7d8',
    'secondary-100': '#bee78a',
    'secondary-200': '#89d329',
    'secondary-300': '#7dc921',
    'secondary-400': '#66b512',
    'secondary-500': '#59a31a',
    'secondary-600': '#3f802a',
    'secondary-700': '#2b6636',
    'secondary-800': '#13532b',
    'secondary-900': '#004422'
};

export const successMap = {
    'success-50': '#d0ffe6',
    'success-100': '#b1ffd5',
    'success-200': '#95fec5',
    'success-300': '#82ffbb',
    'success-400': '#00ff75',
    'success-500': '#00e66a',
    'success-600': '#00d964',
    'success-700': '#01c059',
    'success-800': '#009d48',
    'success-900': '#007536'
};

export const dangerMap = {
    'danger-50': '#ffe3ca',
    'danger-100': '#ffbf84',
    'danger-200': '#ffa148',
    'danger-300': '#ff8700',
    'danger-400': '#ff6f00',
    'danger-500': '#f56700',
    'danger-600': '#f56700',
    'danger-700': '#dd5c00',
    'danger-800': '#a23e00',
    'danger-900': '#6d2900'
};

export const ancillaryMap = {
    'ancillary-100': '#d30f4b',
    'ancillary-200': '#624963',
    'ancillary-300': '#443247'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#1E5474',
    'dataVizComparative-100': '#00BCFF',
    'dataVizComparative-200': '#0D683B',
    'dataVizComparative-300': '#89D329',
    'dataVizComparative-400': '#6A4C72',
    'dataVizComparative-500': '#CBA0D7',
    'dataVizComparative-600': '#980733',
    'dataVizComparative-700': '#DE0043',
    'dataVizComparative-800': '#EAEAEA',
    'dataVizComparative-900': '#5C5C5C'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#092B3E',
    'dataVizSequential1-100': '#10384F',
    'dataVizSequential1-200': '#00617F',
    'dataVizSequential1-300': '#0075A6',
    'dataVizSequential1-400': '#0091DF',
    'dataVizSequential1-500': '#00BCFF',
    'dataVizSequential1-600': '#4FD1FF',
    'dataVizSequential1-700': '#A2E6FF',
    'dataVizSequential1-800': '#C7F0FF',
    'dataVizSequential1-900': '#E8F9FF'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#012B16',
    'dataVizSequential2-100': '#004422',
    'dataVizSequential2-200': '#2B6636',
    'dataVizSequential2-300': '#108000',
    'dataVizSequential2-400': '#66B512',
    'dataVizSequential2-500': '#89D329',
    'dataVizSequential2-600': '#A2DF52',
    'dataVizSequential2-700': '#CBF594',
    'dataVizSequential2-800': '#D8F8AB',
    'dataVizSequential2-900': '#E9F9D4'
};
