// @element/themes/lib/tokens/velocity.theme.js
export const primaryMap = {
    'primary-50': '#daecff',
    'primary-100': '#b3d9ff',
    'primary-200': '#93c9ff',
    'primary-300': '#76baff',
    'primary-400': '#4d9ff2',
    'primary-500': '#338ee7',
    'primary-600': '#1976d2',
    'primary-700': '#0464c4',
    'primary-800': '#0e59a2',
    'primary-900': '#18445d'
};

export const secondaryMap = {
    'secondary-50': '#d5e0ff',
    'secondary-100': '#9eb8ff',
    'secondary-200': '#87a4f5',
    'secondary-300': '#6c8eec',
    'secondary-400': '#4e72d5',
    'secondary-500': '#3056bf',
    'secondary-600': '#2a4ba6',
    'secondary-700': '#293e87',
    'secondary-800': '#1e306e',
    'secondary-900': '#152457'
};

export const dangerMap = {
    'danger-50': '#fbf6ef',
    'danger-100': '#fff3f5',
    'danger-200': '#ee9894',
    'danger-300': '#e3716a',
    'danger-400': '#ec5145',
    'danger-500': '#ef4227',
    'danger-600': '#ce3f09',
    'danger-700': '#cf2d22',
    'danger-800': '#c2261b',
    'danger-900': '#b00320'
};

export const successMap = {
    'success-50': '#eeffef',
    'success-100': '#d3ffd7',
    'success-200': '#a9f6af',
    'success-300': '#6fef7a',
    'success-400': '#29e539',
    'success-500': '#13c545',
    'success-600': '#0ea01a',
    'success-700': '#01830c',
    'success-800': '#075a0e',
    'success-900': '#053709'
};

export const dataVizComparativeMap = {
    'dataVizComparative-50': '#e02020',
    'dataVizComparative-100': '#0091ff',
    'dataVizComparative-200': '#063682',
    'dataVizComparative-300': '#44cc00',
    'dataVizComparative-400': '#e0e0e0',
    'dataVizComparative-500': '#757575',
    'dataVizComparative-600': '#980733',
    'dataVizComparative-700': '#512da8',
    'dataVizComparative-800': '#e65100',
    'dataVizComparative-900': '#89d329'
};

export const dataVizSequential1Map = {
    'dataVizSequential1-50': '#18445d',
    'dataVizSequential1-100': '#0e59a2',
    'dataVizSequential1-200': '#0464c4',
    'dataVizSequential1-300': '#1976d2',
    'dataVizSequential1-400': '#1976d2',
    'dataVizSequential1-500': '#4d9ff2',
    'dataVizSequential1-600': '#76baff',
    'dataVizSequential1-700': '#93c9ff',
    'dataVizSequential1-800': '#b3d9ff',
    'dataVizSequential1-900': '#daecff'
};

export const dataVizSequential2Map = {
    'dataVizSequential2-50': '#152457',
    'dataVizSequential2-100': '#1e306e',
    'dataVizSequential2-200': '#293e87',
    'dataVizSequential2-300': '#2a4ba6',
    'dataVizSequential2-400': '#3056bf',
    'dataVizSequential2-500': '#4e72d5',
    'dataVizSequential2-600': '#6c8eec',
    'dataVizSequential2-700': '#87a4f5',
    'dataVizSequential2-800': '#9eb8ff',
    'dataVizSequential2-900': '#d5e0ff'
};

export const ancillaryMap = {
    'ancillary-100': '#009458',
    'ancillary-200': '#f2c600',
    'ancillary-300': '#32c5ff',
    'ancillary-400': '#6c43f3'
};
