// @element/themes/lib/tokens/verquvo.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    dangerMap,
    successMap,
    lmntThemeDanger,
    lmntThemeDangerVariant,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#D93847';
export const lmntThemePrimaryVariant = ' #910144';
export const lmntThemeSecondary = '#910144';
export const lmntThemeSecondaryVariant = '#910144';

export const primaryMap = {
    'primary-50': '#ff6d7b',
    'primary-100': '#f96573',
    'primary-200': '#f45d6c',
    'primary-300': '#ee5664',
    'primary-400': '#e94e5d',
    'primary-500': '#e34755',
    'primary-600': '#de3f4e',
    'primary-700': '#d93847',
    'primary-800': '#ae1120',
    'primary-900': '#7e010d'
};

export const secondaryMap = {
    'secondary-50': '#c9005d',
    'secondary-100': '#c10059',
    'secondary-200': '#b90056',
    'secondary-300': '#b10052',
    'secondary-400': '#a9004e',
    'secondary-500': '#a1004b',
    'secondary-600': '#990047',
    'secondary-700': '#910144',
    'secondary-800': '#6f0034',
    'secondary-900': '#580029'
};

export const ancillaryMap = {
    'ancillary-100': '#910144'
};
