// @element/themes/lib/tokens/visanne.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#4D4FA8';
export const lmntThemePrimaryVariant = ' #A9CD74';
export const lmntThemeSecondary = '#4D4FA8';
export const lmntThemeSecondaryVariant = '#4D4FA8';
export const lmntThemeDanger = '#CF2D22';
export const lmntThemeDangerVariant = '#A8150B';

export const primaryMap = {
    'primary-50': '#D0D1FF',
    'primary-100': '#A6A8F0',
    'primary-200': '#8B8DE7',
    'primary-300': '#6F71D0',
    'primary-400': '#5F61BE',
    'primary-500': '#4D4FA8',
    'primary-600': '#3D3F9E',
    'primary-700': '#2E3098',
    'primary-800': '#1E208C',
    'primary-900': '#070869'
};

export const secondaryMap = {
    'secondary-50': '#D0D1FF',
    'secondary-100': '#A6A8F0',
    'secondary-200': '#8B8DE7',
    'secondary-300': '#6F71D0',
    'secondary-400': '#5F61BE',
    'secondary-500': '#4D4FA8',
    'secondary-600': '#3D3F9E',
    'secondary-700': '#2E3098',
    'secondary-800': '#1E208C',
    'secondary-900': '#070869'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#A9CD74'
};
