// @element/themes/lib/tokens/vitrakvi.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#10069F';
export const lmntThemePrimaryVariant = ' #F0B323';
export const lmntThemeSecondary = '#A51890';
export const lmntThemeSecondaryVariant = '#A51890';
export const lmntThemeDanger = '#CF2D22';
export const lmntThemeDangerVariant = '#A8150B';

export const primaryMap = {
    'primary-50': '#7B72FA',
    'primary-100': '#6B62ED',
    'primary-200': '#5C53E0',
    'primary-300': '#4D43D3',
    'primary-400': '#3D34C6',
    'primary-500': '#2E24B9',
    'primary-600': '#1F15AC',
    'primary-700': '#10069F',
    'primary-800': '#0C047F',
    'primary-900': '#090360'
};

export const secondaryMap = {
    'secondary-50': '#FD87EB',
    'secondary-100': '#F077DE',
    'secondary-200': '#E367D1',
    'secondary-300': '#D757C4',
    'secondary-400': '#CA47B7',
    'secondary-500': '#BE37AA',
    'secondary-600': '#B1279D',
    'secondary-700': '#A51890',
    'secondary-800': '#8E0C7A',
    'secondary-900': '#770065'
};

export const dangerMap = {
    'danger-50': '#FFD8D6',
    'danger-100': '#FFBAB5',
    'danger-200': '#FDA099',
    'danger-300': '#F3867E',
    'danger-400': '#EA655C',
    'danger-500': '#DF493E',
    'danger-600': '#CF2D22',
    'danger-700': '#BD2218',
    'danger-800': '#A8150B',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#F0B323'
};
