// @element/themes/lib/tokens/xarelto.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    dangerMap,
    successMap,
    lmntThemeDanger,
    lmntThemeDangerVariant,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#D00482';
export const lmntThemePrimaryVariant = '#472783';
export const lmntThemeSecondary = '#472783';
export const lmntThemeSecondaryVariant = '#472783';

export const primaryMap = {
    'primary-50': '#FABFE3',
    'primary-100': '#EB82C3',
    'primary-200': '#E748AA',
    'primary-300': '#E11894',
    'primary-400': '#D00482',
    'primary-500': '#C20279',
    'primary-600': '#B1036F',
    'primary-700': '#A00364',
    'primary-800': '#900159',
    'primary-900': '#720147'
};

export const secondaryMap = {
    'secondary-50': '#C9AEFF',
    'secondary-100': '#AA86F1',
    'secondary-200': '#926AE0',
    'secondary-300': '#7F55CF',
    'secondary-400': '#7146C3',
    'secondary-500': '#623AAE',
    'secondary-600': '#582DA8',
    'secondary-700': '#472783',
    'secondary-800': '#35107A',
    'secondary-900': '#1F0353'
};

export const ancillaryMap = {
    'ancillary-100': '#472783'
};
