// @element/themes/lib/tokens/xofigo.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    dangerMap,
    successMap,
    lmntThemeDanger,
    lmntThemeDangerVariant,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#56004E';
export const lmntThemePrimaryVariant = '#C859A1';
export const lmntThemeSecondary = '#56004E';
export const lmntThemeSecondaryVariant = '#56004E';

export const primaryMap = {
    'primary-50': '#D2B5D0',
    'primary-100': '#BF8CBB',
    'primary-200': '#B769B0',
    'primary-300': '#AD47A4',
    'primary-400': '#A52D9A',
    'primary-500': '#8F0A84',
    'primary-600': '#7B0471',
    'primary-700': '#690360',
    'primary-800': '#56004E',
    'primary-900': '#290226'
};

export const secondaryMap = {
    'secondary-50': '#D2B5D0',
    'secondary-100': '#BF8CBB',
    'secondary-200': '#B769B0',
    'secondary-300': '#AD47A4',
    'secondary-400': '#A52D9A',
    'secondary-500': '#8F0A84',
    'secondary-600': '#7B0471',
    'secondary-700': '#690360',
    'secondary-800': '#56004E',
    'secondary-900': '#290226'
};

export const ancillaryMap = {
    'ancillary-100': '#C859A1'
};
