// @element/themes/lib/tokens/yasmin.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    dangerMap,
    successMap,
    lmntThemeDanger,
    lmntThemeDangerVariant,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#C72439';
export const lmntThemePrimaryVariant = '#E98F1D';
export const lmntThemeSecondary = '#A2007A';
export const lmntThemeSecondaryVariant = '#A2007A';

export const primaryMap = {
    'primary-50': '#FF8998',
    'primary-100': '#F77A8A',
    'primary-200': '#EF6C7C',
    'primary-300': '#E75D6F',
    'primary-400': '#DF4F61',
    'primary-500': '#D74054',
    'primary-600': '#CF3246',
    'primary-700': '#C72439',
    'primary-800': '#AE0118',
    'primary-900': '#8B0012'
};

export const secondaryMap = {
    'secondary-50': '#FF8AE2',
    'secondary-100': '#F176D3',
    'secondary-200': '#E462C4',
    'secondary-300': '#D74EB5',
    'secondary-400': '#C93BA6',
    'secondary-500': '#BC2797',
    'secondary-600': '#AF1388',
    'secondary-700': '#A2007A',
    'secondary-800': '#910076',
    'secondary-900': '#76005C'
};

export const ancillaryMap = {
    'ancillary-100': '#E98F1D'
};
