// @element/themes/lib/tokens/yaz.theme.js
import * as hcp from './hcp.theme';

export const {
    breakpointsMap,
    successMap,
    lmntThemeSuccess,
    lmntThemeSuccessVariant,
    lmntThemeSurface,
    lmntThemeBackground,
    dataVizComparativeMap,
    dataVizSequential1Map,
    dataVizSequential2Map,
    grayMap
} = hcp;

export const lmntThemePrimary = '#347D93';
export const lmntThemePrimaryVariant = ' #347D93';
export const lmntThemeSecondary = '#CE4A0D';
export const lmntThemeSecondaryVariant = '#CE4A0D';
export const lmntThemeDanger = '#B4002B';
export const lmntThemeDangerVariant = '#B4002B';

export const primaryMap = {
    'primary-50': '#C7DFE7',
    'primary-100': '#A0CCD9',
    'primary-200': '#83BDCE',
    'primary-300': '#69ADC1',
    'primary-400': '#53A2BA',
    'primary-500': '#3A8DA6',
    'primary-600': '#347D93',
    'primary-700': '#257086',
    'primary-800': '#1F6377',
    'primary-900': '#0F4A5C'
};

export const secondaryMap = {
    'secondary-50': '#FFCAB0',
    'secondary-100': '#FFB38E',
    'secondary-200': '#FB9B6F',
    'secondary-300': '#F47F47',
    'secondary-400': '#E86629',
    'secondary-500': '#DB5517',
    'secondary-600': '#CE4A0D',
    'secondary-700': '#C2460D',
    'secondary-800': '#AE3C08',
    'secondary-900': '#963204'
};

export const dangerMap = {
    'danger-50': '#FFBDCD',
    'danger-100': '#FF96AF',
    'danger-200': '#F96085',
    'danger-300': '#EB2656',
    'danger-400': '#E2043A',
    'danger-500': '#D30033',
    'danger-600': '#C5002F',
    'danger-700': '#B4002B',
    'danger-800': '#9D0025',
    'danger-900': '#820800'
};

export const ancillaryMap = {
    'ancillary-100': '#347D93'
};
