// @element/react-text-bubble/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime-corejs3/helpers/extends');
var clsx = require('clsx');
var React = require('react');
var PropTypes = require('prop-types');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefault(_extends);
var clsx__default = /*#__PURE__*/_interopDefault(clsx);
var React__default = /*#__PURE__*/_interopDefault(React);
var PropTypes__default = /*#__PURE__*/_interopDefault(PropTypes);

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const TextBubblePropTypes = {
  /**
   * Make the bubble smaller.
   *
   * Defaults to **'default'**.
   */
  bubbleSize: PropTypes__default["default"].oneOf(["default", "small"]),
  /**
   * Optionally override the default component of `div`.
   *
   * Defaults to **'div'**.
   */
  tag: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].elementType]),
  /**
   * Content for the bubble, should be 1 character
   *
   * Defaults to **null**.
   */
  text: PropTypes__default["default"].node,
  /**
   * Color of the badge.
   *
   * Defaults to **'primary'**.
   */
  themeColor: PropTypes__default["default"].oneOf(["primary", "secondary", "danger", "danger-alt", "muted", "blue", "gray", "green", "orange", "purple", "red", "teal", "yellow", "background"]),
  /**
   * The style of the TextBubble.
   *
   * Defaults to **'filled'**.
   */
  variant: PropTypes__default["default"].oneOf(["filled", "outlined"])
};
const TextBubbleDefaultProps = {
  bubbleSize: "default",
  tag: "div",
  text: null,
  themeColor: "primary",
  variant: "filled"
};

const TextBubble = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    tag: CompTag,
    text,
    bubbleSize,
    themeColor,
    children,
    variant,
    ...customProps
  } = props;
  return /*#__PURE__*/React__default["default"].createElement(CompTag, _extends__default["default"]({
    ref: externalRef,
    className: clsx__default["default"]('lmnt lmnt-text-bubble', `lmnt-text-bubble__theme-color--${themeColor}`, variant === 'outlined' && 'lmnt-text-bubble--outlined', bubbleSize === 'small' && 'lmnt-text-bubble--small')
  }, customProps), text, children);
});
TextBubble.displayName = 'TextBubble';
TextBubble.propTypes = TextBubblePropTypes;
TextBubble.defaultProps = TextBubbleDefaultProps;

exports.TextBubble = TextBubble;
