// @element/react-text-bubble/lib/index.es.js
import _extends from '@babel/runtime-corejs3/helpers/extends';
import clsx from 'clsx';
import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const TextBubblePropTypes = {
  /**
   * Make the bubble smaller.
   *
   * Defaults to **'default'**.
   */
  bubbleSize: PropTypes.oneOf(["default", "small"]),
  /**
   * Optionally override the default component of `div`.
   *
   * Defaults to **'div'**.
   */
  tag: PropTypes.oneOfType([PropTypes.string, PropTypes.elementType]),
  /**
   * Content for the bubble, should be 1 character
   *
   * Defaults to **null**.
   */
  text: PropTypes.node,
  /**
   * Color of the badge.
   *
   * Defaults to **'primary'**.
   */
  themeColor: PropTypes.oneOf(["primary", "secondary", "danger", "danger-alt", "muted", "blue", "gray", "green", "orange", "purple", "red", "teal", "yellow", "background"]),
  /**
   * The style of the TextBubble.
   *
   * Defaults to **'filled'**.
   */
  variant: PropTypes.oneOf(["filled", "outlined"])
};
const TextBubbleDefaultProps = {
  bubbleSize: "default",
  tag: "div",
  text: null,
  themeColor: "primary",
  variant: "filled"
};

const TextBubble = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    tag: CompTag,
    text,
    bubbleSize,
    themeColor,
    children,
    variant,
    ...customProps
  } = props;
  return /*#__PURE__*/React.createElement(CompTag, _extends({
    ref: externalRef,
    className: clsx('lmnt lmnt-text-bubble', `lmnt-text-bubble__theme-color--${themeColor}`, variant === 'outlined' && 'lmnt-text-bubble--outlined', bubbleSize === 'small' && 'lmnt-text-bubble--small')
  }, customProps), text, children);
});
TextBubble.displayName = 'TextBubble';
TextBubble.propTypes = TextBubblePropTypes;
TextBubble.defaultProps = TextBubbleDefaultProps;

export { TextBubble };
