// @element/themes/switcher/buildStyleRules.js
const path = require('path');

const buildStyleRules = (config, includePaths, sass, fibers) => {
    config.module.rules.push({
        test: /\.scss$/,
        exclude: /\.lazy\.scss$/i,
        use: [
            'style-loader',
            'css-loader',
            {
                loader: 'sass-loader',
                options: {
                    implementation: sass,
                    sassOptions: {
                        fiber: fibers,
                        includePaths
                    }
                }
            }
        ]
    });

    config.module.rules.push({
        test: /\.lazy\.scss$/i,
        use: [
            {
                loader: 'style-loader',
                options: { injectType: 'lazyStyleTag' }
            },
            'css-loader',
            {
                loader: 'sass-loader',
                options: {
                    implementation: sass,
                    sassOptions: {
                        fiber: fibers,
                        includePaths: [
                            path.resolve(__dirname),
                            path.resolve(__dirname, '../node_modules/@material')
                        ]
                    }
                }
            }
        ]
    });

    return config;
};

module.exports = buildStyleRules;
