// @element/themes/switcher/themesArray.js
module.exports = [
    {
        label: 'Asgrow',
        href: '/asgrow/',
        slug: 'asgrow',
        ancillaryTotal: 1,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '700',
        successVariantIndex: '400',
        dataViz: true
    },
    {
        label: 'Aspirin Cardio',
        href: '/aspirincardio/',
        slug: 'aspirincardio',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Bayer',
        href: '/',
        slug: 'bayer',
        ancillaryTotal: 4,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Bayer Plus Rewards',
        href: '/bayerplusrewards/',
        slug: 'bayerplusrewards',
        ancillaryTotal: 3,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '700',
        successVariantIndex: '900',
        dataViz: true
    },
    {
        label: 'Channel',
        href: '/channel/',
        slug: 'channel',
        ancillaryTotal: 2,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Climate Portal',
        href: '/climateportal/',
        slug: 'climateportal',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '500',
        successVariantIndex: '300',
        dataViz: true
    },
    {
        label: 'Climate FieldView',
        href: '/climatefieldview/',
        slug: 'climatefieldview',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '50',
        successPrimaryIndex: '500',
        successVariantIndex: '50',
        ancillaryTotal: 6,
        dataViz: true
    },
    {
        label: 'DAD',
        href: '/dad/',
        slug: 'dad',
        ancillaryTotal: 3,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Data Enablement',
        href: '/dataenablement/',
        slug: 'dataenablement',
        ancillaryTotal: 2,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        dataViz: true,
        successPrimaryIndex: '600',
        successVariantIndex: '500'
    },
    {
        label: 'Dekalb',
        href: '/dekalb/',
        slug: 'dekalb',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Deltapine',
        href: '/deltapine/',
        slug: 'deltapine',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'DevTools',
        href: '/devtools/',
        slug: 'devtools',
        ancillaryTotal: 3,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Diane 35',
        href: '/diane35/',
        slug: 'diane35',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Element',
        href: '/element/',
        slug: 'element',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Eylea',
        href: '/eylea/',
        slug: 'eylea',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Fontanelle',
        href: '/fontanelle/',
        slug: 'fontanelle',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Gold Country',
        href: '/goldcountry/',
        slug: 'goldcountry',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900'
    },
    {
        label: 'Healthcare Provider Hub',
        href: '/hcp/',
        slug: 'hcp',
        ancillaryTotal: 3,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true
    },
    {
        label: 'Hubner',
        href: '/hubnerseed/',
        slug: 'hubnerseed',
        dangerPrimaryIndex: '400',
        dangerVariantIndex: '700',
        successPrimaryIndex: '600',
        successVariantIndex: '500'
    },
    {
        label: 'Intacta',
        href: '/intacta/',
        slug: 'intacta',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500'
    },
    {
        label: 'Jivi',
        href: '/jivi/',
        slug: 'jivi',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Jung',
        href: '/jungseedgenetics/',
        slug: 'jungseedgenetics',
        dangerPrimaryIndex: '400',
        dangerVariantIndex: '700',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Kerendia',
        href: '/kerendia/',
        slug: 'kerendia',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Kovaltry',
        href: '/kovaltry/',
        slug: 'kovaltry',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Kruger',
        href: '/kruger/',
        slug: 'kruger',
        ancillaryTotal: 3,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500'
    },
    {
        label: 'Kyleena',
        href: '/kyleena/',
        slug: 'kyleena',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'La Tijereta',
        href: '/latijereta/',
        slug: 'latijereta',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500'
    },
    {
        label: 'Lewis',
        href: '/lewishybrids/',
        slug: 'lewishybrids',
        ancillaryTotal: 3,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500'
    },
    {
        label: 'Mirena',
        href: '/mirena/',
        slug: 'mirena',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Monsoy',
        href: '/monsoy/',
        slug: 'monsoy',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Nexavar',
        href: '/nexavar/',
        slug: 'nexavar',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Nuqeba',
        href: '/nuqeba/',
        slug: 'nuqeba',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Qlaira',
        href: '/qlaira/',
        slug: 'qlaira',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'REA',
        href: '/rea/',
        slug: 'rea',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Specialty',
        href: '/specialtyhybrids/',
        slug: 'specialtyhybrids',
        ancillaryTotal: 2,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Stewart',
        href: '/stewart/',
        slug: 'stewart',
        ancillaryTotal: 3,
        dangerPrimaryIndex: '400',
        dangerVariantIndex: '900',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Stivarga',
        href: '/stivarga/',
        slug: 'stivarga',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Stone',
        href: '/stoneseed/',
        slug: 'stoneseed',
        ancillaryTotal: 3,
        dangerPrimaryIndex: '400',
        dangerVariantIndex: '700',
        successPrimaryIndex: '600',
        successVariantIndex: '500'
    },
    {
        label: 'Veg State',
        href: '/vegstate/',
        slug: 'vegstate',
        ancillaryTotal: 3,
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '700',
        successVariantIndex: '900',
        dataViz: true
    },
    {
        label: 'Velocity',
        href: '/velocity/',
        slug: 'velocity',
        ancillaryTotal: 4,
        dangerPrimaryIndex: '900',
        dangerVariantIndex: '600',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true
    },
    {
        label: 'Verquvo',
        href: '/verquvo/',
        slug: 'verquvo',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Visanne',
        href: '/visanne/',
        slug: 'visanne',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Vitrakvi',
        href: '/vitrakvi/',
        slug: 'vitrakvi',
        dangerPrimaryIndex: '600',
        dangerVariantIndex: '800',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Westbred',
        href: '/westbred/',
        slug: 'westbred',
        dangerPrimaryIndex: '400',
        dangerVariantIndex: '700',
        successPrimaryIndex: '600',
        successVariantIndex: '500'
    },
    {
        label: 'Xarelto',
        href: '/xarelto/',
        slug: 'xarelto',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Xofigo',
        href: '/xofigo/',
        slug: 'xofigo',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Yasmin',
        href: '/yasmin/',
        slug: 'yasmin',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '400',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    },
    {
        label: 'Yaz',
        href: '/yaz/',
        slug: 'yaz',
        dangerPrimaryIndex: '700',
        dangerVariantIndex: '700',
        successPrimaryIndex: '600',
        successVariantIndex: '500',
        dataViz: true,
        grayMap: true,
        ancillaryTotal: 1
    }
];
