// @element/react-elevation/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime-corejs3/helpers/extends');
var React = require('react');
var clsx = require('clsx');
var reactCommon = require('@element/react-common');
var PropTypes = require('prop-types');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefault(_extends);
var React__default = /*#__PURE__*/_interopDefault(React);
var clsx__default = /*#__PURE__*/_interopDefault(clsx);
var PropTypes__default = /*#__PURE__*/_interopDefault(PropTypes);

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const ElevationPropTypes = {
  /**
   * Content to be rendered inside the elevation.
   *
   * Defaults to **null**.
   */
  children: PropTypes__default["default"].node,
  /**
   * Adds a custom padding style to the child content. Must be a valid CSS padding declaration. i.e. 38px 30px
   *
   * Defaults to **null**.
   */
  customPadding: PropTypes__default["default"].string,
  /**
   * Sets the distance of the elevation. Value must be between 0-24.
   *
   * Defaults to **0**.
   */
  elevation: PropTypes__default["default"].oneOf([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24]),
  /**
   * Avoids adding an extra element to the DOM and applies proper classes directly to the child element.
   *
   * Defaults to **null**.
   */
  noWrap: PropTypes__default["default"].bool,
  /**
   * The elevation element will have a visible border.
   *
   * Defaults to **false**.
   */
  outlined: PropTypes__default["default"].bool,
  /**
   * Adds a universal padding around the child content. `dense` is 8px, `standard` is 16px, and `airy` is 24px. If `customPadding` is used, this will be ignored.
   *
   * Defaults to **undefined**.
   */
  padding: PropTypes__default["default"].oneOf(["standard", "dense", "airy"]),
  /**
   * The elevation element will have a border radius of 8px.
   *
   * Defaults to **false**.
   */
  rounded: PropTypes__default["default"].bool,
  /**
   * HTML tag to render the contents in.
   *
   * Defaults to **'div'**.
   */
  tag: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].elementType, PropTypes__default["default"].instanceOf(React__default["default"].Component)]),
  /**
   * Adds animation when elevation is changed.
   *
   * Defaults to **null**.
   */
  transition: PropTypes__default["default"].bool
};
const ElevationDefaultProps = {
  children: null,
  customPadding: null,
  elevation: 0,
  noWrap: null,
  outlined: false,
  padding: undefined,
  rounded: false,
  tag: "div",
  transition: null
};

const Elevation = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    children,
    className,
    customPadding,
    elevation,
    noWrap,
    padding,
    rounded,
    tag,
    outlined,
    style,
    transition,
    ...customProps
  } = props;
  const ref = React.useRef();
  React.useImperativeHandle(externalRef, () => ref.current, []);
  const classes = React.useMemo(() => clsx__default["default"]('lmnt', `mdc-elevation--z${elevation}`, padding === 'standard' && `lmnt-elevation--padding-standard`, padding === 'dense' && `lmnt-elevation--padding-dense`, padding === 'airy' && `lmnt-elevation--padding-airy`, transition && 'mdc-elevation-transition', rounded && 'lmnt-elevation--rounded', outlined && 'lmnt-elevation--outlined', className), [elevation, padding, transition, rounded, outlined, className]);
  const styles = React.useMemo(() => ({
    ...style,
    padding: customPadding
  }), [style, customPadding]);
  const Tag = React.useMemo(() => tag, [tag]);
  return (
    /*#__PURE__*/
    // eslint-disable-next-line react/jsx-no-useless-fragment
    React__default["default"].createElement(React__default["default"].Fragment, null, noWrap ? reactCommon.addProps(children, {
      className: classes,
      ...customProps
    }) : /*#__PURE__*/React__default["default"].createElement(Tag, _extends__default["default"]({
      ref: ref,
      className: classes,
      style: styles
    }, customProps), children))
  );
});
Elevation.displayName = 'Elevation';
Elevation.propTypes = ElevationPropTypes;
Elevation.defaultProps = ElevationDefaultProps;
var Elevation$1 = /*#__PURE__*/React__default["default"].memo(Elevation);

exports.Elevation = Elevation$1;
