// @element/react-padding/lib/index.es.js
import _extends from '@babel/runtime-corejs3/helpers/extends';
import React, { forwardRef, useMemo, useRef, useImperativeHandle } from 'react';
import clsx from 'clsx';
import PropTypes from 'prop-types';

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const PaddingPropTypes = {
  /**
   * Content to be rendered inside the padding component. Accepts any valid markup.
   *
   * Defaults to **null**.
   */
  children: PropTypes.node.isRequired,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes.string,
  /**
   * Adds a custom padding style to the child content. Must be a valid CSS padding declaration. i.e. 38px 30px
   *
   * Defaults to **null**.
   */
  customPadding: PropTypes.string,
  /**
   * By default a div will be rendered, this allow another tag to be used instead. A block type component is suggested.
   *
   * Defaults to **'div'**.
   */
  tag: PropTypes.oneOfType([PropTypes.string, PropTypes.elementType]),
  /**
   * Applies the specified padding amount. `dense` is 8px, `standard` is 16px, and `airy` is 24px. If `customPadding` is used, this will be ignored.
   *
   * Defaults to **'standard'**.
   */
  variant: PropTypes.oneOf(["dense", "standard", "airy"])
};
const PaddingDefaultProps = {
  children: null,
  className: undefined,
  customPadding: null,
  tag: "div",
  variant: "standard"
};

const Padding = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    tag,
    variant,
    children,
    className,
    customPadding,
    style,
    ...customProps
  } = props;
  const styles = useMemo(() => ({
    ...style,
    padding: customPadding
  }), [style, customPadding]);
  const Tag = tag || 'div';
  const ref = useRef();
  useImperativeHandle(externalRef, () => ref.current, []);
  return /*#__PURE__*/React.createElement(Tag, _extends({
    className: clsx('lmnt', 'lmnt-padding', variant === 'dense' && 'lmnt-padding--dense', variant === 'standard' && 'lmnt-padding--standard', variant === 'airy' && 'lmnt-padding--airy', className),
    style: styles
  }, customProps), children);
});
Padding.displayName = 'Padding';
Padding.propTypes = PaddingPropTypes;
Padding.defaultProps = PaddingDefaultProps;

export { Padding };
