// @element/react-padding/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime-corejs3/helpers/extends');
var React = require('react');
var clsx = require('clsx');
var PropTypes = require('prop-types');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefault(_extends);
var React__default = /*#__PURE__*/_interopDefault(React);
var clsx__default = /*#__PURE__*/_interopDefault(clsx);
var PropTypes__default = /*#__PURE__*/_interopDefault(PropTypes);

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const PaddingPropTypes = {
  /**
   * Content to be rendered inside the padding component. Accepts any valid markup.
   *
   * Defaults to **null**.
   */
  children: PropTypes__default["default"].node.isRequired,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes__default["default"].string,
  /**
   * Adds a custom padding style to the child content. Must be a valid CSS padding declaration. i.e. 38px 30px
   *
   * Defaults to **null**.
   */
  customPadding: PropTypes__default["default"].string,
  /**
   * By default a div will be rendered, this allow another tag to be used instead. A block type component is suggested.
   *
   * Defaults to **'div'**.
   */
  tag: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].elementType]),
  /**
   * Applies the specified padding amount. `dense` is 8px, `standard` is 16px, and `airy` is 24px. If `customPadding` is used, this will be ignored.
   *
   * Defaults to **'standard'**.
   */
  variant: PropTypes__default["default"].oneOf(["dense", "standard", "airy"])
};
const PaddingDefaultProps = {
  children: null,
  className: undefined,
  customPadding: null,
  tag: "div",
  variant: "standard"
};

const Padding = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    tag,
    variant,
    children,
    className,
    customPadding,
    style,
    ...customProps
  } = props;
  const styles = React.useMemo(() => ({
    ...style,
    padding: customPadding
  }), [style, customPadding]);
  const Tag = tag || 'div';
  const ref = React.useRef();
  React.useImperativeHandle(externalRef, () => ref.current, []);
  return /*#__PURE__*/React__default["default"].createElement(Tag, _extends__default["default"]({
    className: clsx__default["default"]('lmnt', 'lmnt-padding', variant === 'dense' && 'lmnt-padding--dense', variant === 'standard' && 'lmnt-padding--standard', variant === 'airy' && 'lmnt-padding--airy', className),
    style: styles
  }, customProps), children);
});
Padding.displayName = 'Padding';
Padding.propTypes = PaddingPropTypes;
Padding.defaultProps = PaddingDefaultProps;

exports.Padding = Padding;
