// @monsantoit/profile-client/src/commonQueries.js
/*
 * decaffeinate suggestions:
 * DS102: Remove unnecessary code created because of implicit returns
 * Full docs: https://github.com/decaffeinate/decaffeinate/blob/master/docs/suggestions.md
 */

module.exports = {
    currentUser() {
        return () => ({
            query: `
                { 
                    getCurrentUser { 
                        id 
                        preferredName {first, last, middle, full} 
                        email 
                        employeeType 
                    } 
                }`,
            variables: {}
        })
    },
    currentUserWithEntitlements(appId) {
        return () => ({
            query: `
                query ($appId: String!) { 
                    getCurrentUser { 
                        id 
                        preferredName {first, last, middle, full} 
                        entitlements(appId: $appId) { 
                            code 
                        } 
                    } 
                }`,
            variables: {appId}
        })
    },
    usersByIdWithEntitlements(userIds, appId) {
        return () => ({
            query: `
                query ($userIds: [String!]!, $appId: String!) { 
                    getUsersById(ids: $userIds) { 
                        id 
                        preferredName {first, last, middle, full} 
                        entitlements(appId: $appId) { code } 
                    } 
                }`,
            variables: {userIds, appId}
        })
    }
}
