// @monsantoit/profile-client/lib/commonQueries.js
"use strict";

/*
 * decaffeinate suggestions:
 * DS102: Remove unnecessary code created because of implicit returns
 * Full docs: https://github.com/decaffeinate/decaffeinate/blob/master/docs/suggestions.md
 */

module.exports = {
  currentUser: function currentUser() {
    return function () {
      return {
        query: "\n                { \n                    getCurrentUser { \n                        id \n                        preferredName {first, last, middle, full} \n                        email \n                        employeeType \n                    } \n                }",
        variables: {}
      };
    };
  },
  currentUserWithEntitlements: function currentUserWithEntitlements(appId) {
    return function () {
      return {
        query: "\n                query ($appId: String!) { \n                    getCurrentUser { \n                        id \n                        preferredName {first, last, middle, full} \n                        entitlements(appId: $appId) { \n                            code \n                        } \n                    } \n                }",
        variables: {
          appId: appId
        }
      };
    };
  },
  usersByIdWithEntitlements: function usersByIdWithEntitlements(userIds, appId) {
    return function () {
      return {
        query: "\n                query ($userIds: [String!]!, $appId: String!) { \n                    getUsersById(ids: $userIds) { \n                        id \n                        preferredName {first, last, middle, full} \n                        entitlements(appId: $appId) { code } \n                    } \n                }",
        variables: {
          userIds: userIds,
          appId: appId
        }
      };
    };
  }
};
