// @element/react-grid/lib/index.es.js
import _extends from '@babel/runtime-corejs3/helpers/extends';
import React, { createContext, forwardRef, useMemo, useRef, useImperativeHandle, useContext } from 'react';
import clsx from 'clsx';
import PropTypes from 'prop-types';
import { deprecatedProp } from '@element/react-common';

const GridContext = /*#__PURE__*/createContext({});

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const GridPropTypes = {
  /**
   * @deprecated _Use `variant`='airy' instead._
   *
   * Applies extra padding to each column.
   *
   * Defaults to **undefined**.
   */
  airy: deprecatedProp(PropTypes.bool, "Grid", 'Use `variant`="airy" instead.'),
  /**
   * Accepts GridCol or GridRow.
   *
   * Defaults to **null**.
   */
  children: PropTypes.node.isRequired,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes.string,
  /**
   * Sets the column gap for the grid.
   *
   * Defaults to **'16px'**.
   */
  columnGap: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Applies 100% height to the grid and direct children rows. Useful when creating a page layout. Be aware all direct children rows will have a 100% height.
   *
   * Defaults to **null**.
   */
  fullHeight: PropTypes.bool,
  /**
   * Applies 100% width to the grid and direct children rows. Useful to maintain full width in a flex container.
   *
   * Defaults to **null**.
   */
  fullWidth: PropTypes.bool,
  /**
   * Alignment of entire grid: left and right.
   *
   * Defaults to **'center'**.
   */
  gridAlign: PropTypes.oneOf(["center", "left", "right"]),
  /**
   * Sets the row gap for the grid.
   *
   * Defaults to **'16px'**.
   */
  rowGap: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Styles for the grid.
   *
   * Defaults to **undefined**.
   */
  style: PropTypes.object,
  /**
   * If true the grid will have a max width according to your theme's breakpoints.
   *
   * Defaults to **false**.
   */
  useMaxWidth: PropTypes.bool,
  /**
   * Select `airy` or `standard` padding.
   *
   * Defaults to **'standard'**.
   */
  variant: PropTypes.oneOf(["standard", "airy", "none"])
};
const GridDefaultProps = {
  airy: undefined,
  children: null,
  className: undefined,
  columnGap: "16px",
  fullHeight: null,
  fullWidth: null,
  gridAlign: "center",
  rowGap: "16px",
  style: undefined,
  useMaxWidth: false,
  variant: "standard"
};

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const GridColPropTypes = {
  /**
   * Alignment of cell within the row: top, middle, bottom. If omitted the col will stretch to the height of the row.
   *
   * Defaults to **undefined**.
   */
  align: PropTypes.oneOf(["", "top", "middle", "bottom"]),
  /**
   * Content to be rendered inside the GridCol. Accepts any valid markup.
   *
   * Defaults to **null**.
   */
  children: PropTypes.node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes.string,
  /**
   * Specifies the number of columns the cell spans.
   *
   * Defaults to **null**.
   */
  desktopCol: PropTypes.number,
  /**
   * Specifies a grid item's start position within the grid column by contributing a line, a span, or nothing (automatic) to its grid placement. Set to any valid `grid-column-start`.
   *
   * Defaults to **null**.
   */
  gridColStart: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Horizontal alignment of the cell contents of cell content: top, middle, bottom.
   *
   * Defaults to **'left'**.
   */
  horizontalAlign: PropTypes.oneOf(["left", "center", "right"]),
  /**
   * Specifies the order of the cell.
   *
   * Defaults to **null**.
   */
  order: PropTypes.number,
  /**
   * Specifies the number of columns the cell spans on a phone.
   *
   * Defaults to **null**.
   */
  phoneCol: PropTypes.number,
  /**
   * Specifies the number of columns the cell spans on a tablet.
   *
   * Defaults to **null**.
   */
  tabletCol: PropTypes.number,
  /**
   * Vertical alignment of cell content: top, middle, bottom.
   *
   * Defaults to **'top'**.
   */
  verticalAlign: PropTypes.oneOf(["top", "middle", "bottom"])
};
const GridColDefaultProps = {
  align: undefined,
  children: null,
  className: undefined,
  desktopCol: null,
  gridColStart: null,
  horizontalAlign: "left",
  order: null,
  phoneCol: null,
  tabletCol: null,
  verticalAlign: "top"
};

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const GridRowPropTypes = {
  /**
   * Content to be rendered inside the GridRow. Accepts any valid markup.
   *
   * Defaults to **null**.
   */
  children: PropTypes.node.isRequired,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes.string
};
const GridRowDefaultProps = {
  children: null,
  className: undefined
};

const Grid = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    airy,
    children,
    className,
    columnGap,
    fullHeight,
    fullWidth,
    gridAlign,
    rowGap,
    style,
    useMaxWidth,
    variant,
    ...customProps
  } = props;
  const currentVariant = airy ? 'airy' : variant;
  const value = useMemo(() => ({
    fullHeight,
    rowGap,
    columnGap
  }), [columnGap, fullHeight, rowGap]);
  const ref = useRef();
  useImperativeHandle(externalRef, () => ref.current, []);
  const classes = useMemo(() => clsx('lmnt', 'mdc-layout-grid', 'lmnt-layout-grid', `mdc-layout-grid mdc-layout-grid--align-${gridAlign}`, (currentVariant === 'airy' || airy) && 'mdc-grid-cell-wrapper-airy-tablet mdc-grid-cell-wrapper-airy-desktop', variant === 'none' && 'mdc-grid-cell-wrapper-none-tablet mdc-grid-cell-wrapper-none-desktop mdc-grid-cell-wrapper-none-phone ', fullWidth && 'lmnt-layout-grid--full-width', fullHeight && 'lmnt-layout-grid--full-height', useMaxWidth && 'lmnt-layout-grid--with-max-width', className), [airy, className, currentVariant, fullHeight, fullWidth, gridAlign, useMaxWidth, variant]);
  return /*#__PURE__*/React.createElement(GridContext.Provider, {
    value: value
  }, /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: classes
  }, customProps), children));
});
Grid.displayName = 'Grid';
Grid.propTypes = GridPropTypes;
Grid.defaultProps = GridDefaultProps;

const GridCol = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    align,
    children,
    className,
    columns,
    desktopCol,
    gridColStart,
    horizontalAlign,
    order,
    phoneCol,
    style,
    tabletCol,
    verticalAlign,
    ...customProps
  } = props;
  const styles = useMemo(() => ({
    ...style,
    gridColumnStart: gridColStart
  }), [gridColStart, style]);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: externalRef,
    style: styles,
    className: clsx('lmnt', `lmnt-layout-grid__cell--vertical-align-${verticalAlign}`, `lmnt-layout-grid__cell--horizontal-align-${horizontalAlign}`, 'mdc-layout-grid__cell', !columns && !phoneCol && !tabletCol && !desktopCol && [`mdc-layout-grid__cell`], order && `mdc-layout-grid__cell--order-${order}`, align && `mdc-layout-grid__cell--align-${align}`, columns && [`mdc-layout-grid__cell--span-${columns}`], phoneCol && [`mdc-layout-grid__cell--span-${phoneCol}-phone`], tabletCol && [`mdc-layout-grid__cell--span-${tabletCol}-tablet`], desktopCol && [`mdc-layout-grid__cell--span-${desktopCol}-desktop`], 'lmnt-layout-grid__cell', className)
  }, customProps), children);
});
GridCol.displayName = 'GridCol';
GridCol.propTypes = GridColPropTypes;
GridCol.defaultProps = GridColDefaultProps;

const GridRow = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    children,
    className,
    style,
    ...customProps
  } = props;
  const {
    fullHeight,
    rowGap,
    columnGap
  } = useContext(GridContext);
  const styles = useMemo(() => ({
    ...style,
    height: fullHeight ? '100%' : '',
    rowGap,
    columnGap
  }), [style, fullHeight, rowGap, columnGap]);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: externalRef,
    className: clsx('lmnt', 'mdc-layout-grid__inner', 'lmnt-layout-grid__inner', className),
    style: styles
  }, customProps), children);
});
GridRow.displayName = 'GridRow';
GridRow.propTypes = GridRowPropTypes;
GridRow.defaultProps = GridRowDefaultProps;

export { Grid, GridCol, GridContext, GridRow };
