// @element/react-floating-label/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime-corejs3/helpers/extends');
var floatingLabel = require('@material/floating-label');
var clsx = require('clsx');
var React = require('react');
var PropTypes = require('prop-types');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefault(_extends);
var clsx__default = /*#__PURE__*/_interopDefault(clsx);
var React__default = /*#__PURE__*/_interopDefault(React);
var PropTypes__default = /*#__PURE__*/_interopDefault(PropTypes);

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const FloatingLabelPropTypes = {
  /**
   * Content to be rendered inside the label. Expects a string, though accepts any valid markup.
   *
   * Defaults to **null**.
   */
  children: PropTypes__default["default"].node.isRequired,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes__default["default"].string,
  /**
   * The label will appear in the floating state.
   *
   * Defaults to **false**.
   */
  float: PropTypes__default["default"].bool,
  /**
   * The id of the component to trigger on active
   *
   * Defaults to **empty string**.
   */
  htmlFor: PropTypes__default["default"].string,
  /**
   * The label will shake briefly when set.
   *
   * Defaults to **false**.
   */
  shake: PropTypes__default["default"].bool
};
const FloatingLabelDefaultProps = {
  children: null,
  className: undefined,
  float: false,
  htmlFor: "",
  shake: false
};

const FloatingLabel = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    children,
    htmlFor = '',
    shake = false,
    float = false,
    className,
    style,
    ...customProps
  } = props;
  const mdcFloatingLabelRef = React.useRef();
  const content = children;
  const ref = React.useRef();
  React.useImperativeHandle(externalRef, () => ref.current, []);
  React.useEffect(() => {
    mdcFloatingLabelRef.current = floatingLabel.MDCFloatingLabel.attachTo(ref.current);
    return () => {
      mdcFloatingLabelRef.current.destroy();
    };
  });
  React.useEffect(() => {
    mdcFloatingLabelRef.current.float(float);
  }, [float]);
  return /*#__PURE__*/React__default["default"].createElement("label", _extends__default["default"]({
    ref: ref,
    htmlFor: htmlFor,
    className: clsx__default["default"]('lmnt', 'mdc-floating-label', float && 'mdc-floating-label--float-above', shake && 'mdc-floating-label--shake', className)
  }, customProps, {
    style: style
  }), content);
});
FloatingLabel.displayName = 'FloatingLabel';
FloatingLabel.propTypes = FloatingLabelPropTypes;
FloatingLabel.defaultProps = FloatingLabelDefaultProps;

exports.FloatingLabel = FloatingLabel;
