// @monsantoit/phoenix-html-middleware/config/store.js
const {Config, source, processor} = require('@monsantoit/config')
const {determineEnvironmentFileName, isFargate} = require('@monsantoit/config-velocity-store')

const environmentFileName =
    process.env.NODE_ENV === 'production' && isFargate
        ? `config.${determineEnvironmentFileName()}`
        : `config.${process.env.CONFIG_ENV || process.env.NODE_ENV || 'default'}.js`
console.log('environmentFileName - ', environmentFileName)
const config = new Config({
    sources: [source.fromJS({src: `${__dirname}/${environmentFileName}`})],
    processors: [
        processor.readVaultFromConfig({
            enabled: true,
            auth: {
                type: 'auto',
                roleName: {inConfig: 'profile-approle.role_name'}, // prefer this for np and prod
                // roleId/secretId needed for CI/CD
                roleId: {inConfig: 'profile-approle.role_id'},
                secretId: {inConfig: 'profile-approle.secret_id'},
            },
        }),
    ],
    required: [],
})

module.exports = config
