// @element/react-notched-outline/lib/index.es.js
import _extends from '@babel/runtime-corejs3/helpers/extends';
import React, { forwardRef, useRef, useImperativeHandle, useEffect } from 'react';
import { MDCNotchedOutline } from '@material/notched-outline';
import clsx from 'clsx';
import PropTypes from 'prop-types';

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const NotchedOutlinePropTypes = {
  /**
   * Content to be rendered inside the notched outline. Accepts a string.
   *
   * Defaults to **null**.
   */
  children: PropTypes.node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes.string,
  /**
   * Determines if the notch is opened or closed.
   *
   * Defaults to **undefined**.
   */
  closed: PropTypes.bool,
  /**
   *
   *
   * Defaults to **null**.
   */
  width: PropTypes.number
};
const NotchedOutlineDefaultProps = {
  children: null,
  className: undefined,
  closed: undefined,
  width: null
};

const NotchedOutline = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    closed,
    className,
    children,
    width,
    ...customProps
  } = props;
  const ref = useRef();
  const mdcNotchedOutlineRef = useRef();
  useImperativeHandle(externalRef, () => ref.current, []);
  useEffect(() => {
    mdcNotchedOutlineRef.current = MDCNotchedOutline.attachTo(ref.current);
    return () => {
      mdcNotchedOutlineRef.current.destroy();
    };
  }, []);
  useEffect(() => {
    if (closed === undefined) return;
    closed ? mdcNotchedOutlineRef.current.closeNotch() : mdcNotchedOutlineRef.current.notch();
  }, [closed]);
  useEffect(() => {
    if (width) {
      mdcNotchedOutlineRef.current.foundation.adapter.setNotchWidthProperty(width);
    } else {
      mdcNotchedOutlineRef.current.foundation.adapter.removeNotchWidthProperty();
    }
  }, [width]);
  return /*#__PURE__*/React.createElement("div", _extends({
    className: clsx('lmnt', 'mdc-notched-outline', className),
    ref: ref
  }, customProps), /*#__PURE__*/React.createElement("div", {
    className: "mdc-notched-outline__leading"
  }), /*#__PURE__*/React.createElement("div", {
    className: "mdc-notched-outline__notch"
  }, children), /*#__PURE__*/React.createElement("div", {
    className: "mdc-notched-outline__trailing"
  }));
});
NotchedOutline.displayName = 'NotchedOutline';
NotchedOutline.propTypes = NotchedOutlinePropTypes;
NotchedOutline.defaultProps = NotchedOutlineDefaultProps;

export { NotchedOutline };
