// @element/react-line-ripple/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime-corejs3/helpers/extends');
var React = require('react');
var lineRipple = require('@material/line-ripple');
var clsx = require('clsx');
var PropTypes = require('prop-types');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefault(_extends);
var React__default = /*#__PURE__*/_interopDefault(React);
var clsx__default = /*#__PURE__*/_interopDefault(clsx);
var PropTypes__default = /*#__PURE__*/_interopDefault(PropTypes);

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const LineRipplePropTypes = {
  /**
   *
   *
   * Defaults to **null**.
   */
  activated: PropTypes__default["default"].bool,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes__default["default"].string,
  /**
   *
   *
   * Defaults to **null**.
   */
  rippleCenter: PropTypes__default["default"].string
};
const LineRippleDefaultProps = {
  activated: null,
  className: undefined,
  rippleCenter: null
};

const LineRipple = /*#__PURE__*/React.forwardRef((props, externalRef) => {
  const {
    activated,
    rippleCenter,
    className,
    ...customProps
  } = props;
  const mdcLineRippleRef = React.useRef();
  const ref = React.useRef();
  React.useImperativeHandle(externalRef, () => ref.current, []);
  React.useEffect(() => {
    mdcLineRippleRef.current = lineRipple.MDCLineRipple.attachTo(ref.current);
    return () => {
      mdcLineRippleRef.current.destroy();
    };
  }, [ref]);
  React.useEffect(() => {
    if (activated) {
      mdcLineRippleRef.current.activate();
    } else {
      mdcLineRippleRef.current.deactivate();
    }
  }, [activated]);
  React.useEffect(() => {
    if (rippleCenter) {
      mdcLineRippleRef.current.setRippleCenter(rippleCenter);
    }
  }, [rippleCenter]);
  return /*#__PURE__*/React__default["default"].createElement("div", _extends__default["default"]({
    ref: ref,
    className: clsx__default["default"]('lmnt', 'mdc-line-ripple', className)
  }, customProps));
});
LineRipple.displayName = 'LineRipple';
LineRipple.propTypes = LineRipplePropTypes;
LineRipple.defaultProps = LineRippleDefaultProps;

exports.LineRipple = LineRipple;
