// @monsantoit/profile-client/lib/safe/browser.js
"use strict";

var browser = {
  window: function (_window) {
    function window() {
      return _window.apply(this, arguments);
    }
    window.toString = function () {
      return _window.toString();
    };
    return window;
  }(function () {
    if (typeof window === 'undefined') {
      return {};
    }
    return window;
  }),
  document: function (_document) {
    function document() {
      return _document.apply(this, arguments);
    }
    document.toString = function () {
      return _document.toString();
    };
    return document;
  }(function () {
    if (typeof document === 'undefined') {
      return {};
    }
    return document;
  }),
  namespace: function namespace() {
    var namespace = browser.window().velocity || browser.window().phoenix;
    if (namespace != null) {
      return namespace;
    }
    console.warn('Could not find the namespace for profile client. Are you using @monsantoit/service-bindings middleware?');
    return {};
  },
  serviceBindings: function serviceBindings() {
    return browser.namespace().serviceBindings || {};
  },
  localDevelopment: function localDevelopment() {
    return !!browser.namespace().localDevelopment;
  },
  localOcelot: function localOcelot() {
    return !!browser.namespace().localOcelot;
  }
};
module.exports = browser;
