// @element/react-hero/lib/index.es.js
import _extends from '@babel/runtime-corejs3/helpers/extends';
import clsx from 'clsx';
import React, { forwardRef, useRef, useImperativeHandle, useMemo } from 'react';
import PropTypes from 'prop-types';
import { Group } from '@element/react-group';
import { Typography } from '@element/react-typography';

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const HeroPropTypes = {
  /**
   * Aligns the Hero contents.
   *
   * Defaults to **'centered'**.
   */
  alignment: PropTypes.oneOf(["center-bottom", "center-top", "centered", "left-bottom", "left-center", "left-top", "right-bottom", "right-center", "right-top"]),
  /**
   * Content to be rendered inside the Hero.
   *
   * Defaults to **null**.
   */
  children: PropTypes.node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes.string,
  /**
   * Image Url for Hero background image
   *
   * Defaults to **null**.
   */
  imageUrl: PropTypes.string,
  /**
   * Adds an overlay over the Hero. Use `overlayProps` to send props to the Hero Overlay component.  See Hero Overlay for prop details.
   *
   * Defaults to **false**.
   */
  overlay: PropTypes.bool,
  /**
   * Props sent to Hero Overlay. See Hero Overlay for prop details.
   *
   * Defaults to **{}**.
   */
  overlayProps: PropTypes.object,
  /**
   * Override the default background color. Can only be used if background image is omitted.
   *
   * Defaults to **empty string**.
   */
  themeColor: PropTypes.oneOf(["", "surface", "primary", "secondary"])
};
const HeroDefaultProps = {
  alignment: "centered",
  children: null,
  className: undefined,
  imageUrl: null,
  overlay: false,
  overlayProps: {},
  themeColor: ""
};

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const HeroContentPropTypes = {
  /**
   * Content to be rendered inside the Hero Content.  Will take precedence over props.
   *
   * Defaults to **null**.
   */
  children: PropTypes.node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes.string,
  /**
   * Button to be rendered inside the Hero Content.
   *
   * Defaults to **null**.
   */
  heroBtn: PropTypes.node,
  /**
   * Logo to be rendered inside the Hero Content.
   *
   * Defaults to **null**.
   */
  logo: PropTypes.node,
  /**
   * Overline text for the Hero Content. Cannot be used with children.
   *
   * Defaults to **'Overline'**.
   */
  overline: PropTypes.string,
  /**
   * Adds a padding style to the Hero. Must be a valid CSS padding declaration. i.e. 38px 30px
   *
   * Defaults to **null**.
   */
  padding: PropTypes.string,
  /**
   * Passthrough style object.
   *
   * Defaults to **null**.
   */
  style: PropTypes.object,
  /**
   * Subtitle text for the Hero Content. Cannot be used with children.
   *
   * Defaults to **'Subtitle'**.
   */
  subtitle: PropTypes.string,
  /**
   * Set the color of the text.  Make sure to check for accessibility compliance.
   *
   * Defaults to **'on-unknown-black-active'**.
   */
  textColor: PropTypes.oneOf(["on-unknown-black-active", "on-unknown-white-active", "primary", "secondary", "on-primary", "on-secondary"]),
  /**
   * Title text for the Hero Content. Cannot be used with children.
   *
   * Defaults to **'Title'**.
   */
  title: PropTypes.string
};
const HeroContentDefaultProps = {
  children: null,
  className: undefined,
  heroBtn: null,
  logo: null,
  overline: "Overline",
  padding: null,
  style: null,
  subtitle: "Subtitle",
  textColor: "on-unknown-black-active",
  title: "Title"
};

/**
 *
 * This file is automatically generated by the Element
 * documentation system. Any modifications to this file
 * will be overwritten!
 *
 */

const HeroOverlayPropTypes = {
  /**
   * Content to be rendered inside the Hero Overlay.  Useful for angles or other non semantic elements displayed in the Hero.
   *
   * Defaults to **null**.
   */
  children: PropTypes.node,
  /**
   * The css class name to be passed through to the component markup.
   *
   * Defaults to **undefined**.
   */
  className: PropTypes.string,
  /**
   * Overlay theme color.
   *
   * Defaults to **'dark'**.
   */
  overlayThemeColor: PropTypes.oneOf(["dark", "light", "surface", "linear-gradient-white-to-bottom", "linear-gradient-white-to-top", "linear-gradient-black-to-bottom", "linear-gradient-black-to-top", "linear-gradient-primary-dark-to-bottom", "linear-gradient-primary-dark-to-top", "linear-gradient-primary-darker-to-bottom", "linear-gradient-primary-darker-to-top", "linear-gradient-primary-to-bottom", "linear-gradient-primary-to-top", "linear-gradient-secondary-dark-to-bottom", "linear-gradient-secondary-dark-to-top", "linear-gradient-secondary-darker-to-bottom", "linear-gradient-secondary-darker-to-top", "linear-gradient-secondary-to-bottom", "linear-gradient-secondary-to-top"]),
  /**
   * Position of the Overlay. Use `overlayThemeColor` to change the theme color. If set to leading or trailing be sure to set `alignment` on Hero accordingly to have Hero Content over the overlay.
   *
   * Defaults to **'full'**.
   */
  position: PropTypes.oneOf(["full", "leading", "trailing"])
};
const HeroOverlayDefaultProps = {
  children: null,
  className: undefined,
  overlayThemeColor: "dark",
  position: "full"
};

const HeroOverlay = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    children,
    className,
    position,
    overlayThemeColor,
    ...customProps
  } = props;
  const ref = useRef();
  useImperativeHandle(externalRef, () => ref.current, []);
  const classNames = useMemo(() => position !== 'none' && clsx('lmnt', 'lmnt-hero-overlay ', !overlayThemeColor.includes('gradient') && overlayThemeColor !== 'surface' && `lmnt-hero-overlay__${overlayThemeColor}`, overlayThemeColor.includes('gradient') && overlayThemeColor !== 'surface' && `lmnt-theme--bg-${overlayThemeColor}`, overlayThemeColor === 'surface' && 'lmnt-theme-surface-bg', `lmnt-hero-overlay__${position}`, className), [overlayThemeColor, position, className]);
  return /*#__PURE__*/React.createElement("div", _extends({
    className: classNames
  }, customProps), children && children);
});
HeroOverlay.displayName = 'HeroOverlay';
HeroOverlay.propTypes = HeroOverlayPropTypes;
HeroOverlay.defaultProps = HeroOverlayDefaultProps;

const Hero = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    alignment,
    children,
    className,
    imageUrl,
    style,
    themeColor,
    overlay,
    overlayProps,
    ...customProps
  } = props;
  const ref = useRef();
  useImperativeHandle(externalRef, () => ref.current, []);
  const classNames = useMemo(() => clsx('lmnt lmnt-hero', themeColor && `lmnt-theme-${themeColor}-bg`, `lmnt-hero__align-${alignment}`, className), [alignment, className, themeColor]);
  const styles = useMemo(() => {
    if (!imageUrl) return style;
    return {
      ...style,
      backgroundImage: `url(${imageUrl})`
    };
  }, [imageUrl, style]);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: classNames,
    style: styles
  }, customProps), overlay && /*#__PURE__*/React.createElement(HeroOverlay, overlayProps), children);
});
Hero.displayName = 'Hero';
Hero.propTypes = HeroPropTypes;
Hero.defaultProps = HeroDefaultProps;

const HeroContent = /*#__PURE__*/forwardRef((props, externalRef) => {
  const {
    children,
    className,
    heroBtn,
    logo,
    overline,
    padding,
    title,
    style,
    subtitle,
    textColor,
    ...customProps
  } = props;
  const ref = useRef();
  const classNames = useMemo(() => clsx('lmnt lmnt-hero-content', `lmnt-theme-${textColor}`, className), [className, textColor]);
  useImperativeHandle(externalRef, () => ref.current, []);
  const styles = useMemo(() => ({
    ...style,
    padding
  }), [padding, style]);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: classNames
  }, customProps, {
    style: styles
  }), children || /*#__PURE__*/React.createElement(Group, {
    gap: "dense",
    direction: "vertical",
    secondaryAlign: "center"
  }, logo && logo, overline && /*#__PURE__*/React.createElement(Typography, {
    type: "overline"
  }, overline), title && /*#__PURE__*/React.createElement(Typography, {
    type: "display2"
  }, title), subtitle && /*#__PURE__*/React.createElement(Typography, {
    type: "display6"
  }, subtitle), heroBtn && heroBtn));
});
HeroContent.displayName = 'HeroContent';
HeroContent.propTypes = HeroContentPropTypes;
HeroContent.defaultProps = HeroContentDefaultProps;

export { Hero, HeroContent, HeroOverlay };
